{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PIC-SURE API tutorial using the Undiagnosed Diseases Network (UDN) database\n",
    "This is a tutorial notebook, aimed to be quickly up and running with the R PIC-SURE API. It covers the main functionalities of the API.\n",
    "\n",
    "## R PIC-SURE API\n",
    "### What is PIC-SURE?\n",
    "Databases exposed through the PIC-SURE API encompass a wide heterogeneity of architectures and data organizations underneath. PIC-SURE hides this complexity and exposes the different databases in the same format, allowing researchers to focus on the analysis and medical insights, thus easing the process of reproducible sciences.\n",
    "\n",
    "### More about PIC-SURE\n",
    "PIC-SURE stands for Patient-centered Information Commons: Standardized Unification of Research Elements. The API is available in two different programming languages, Python and R, allowing investigators to query databases in the same way using either of those languages.\n",
    "\n",
    "PIC-SURE is a larger project from which the R/python PIC-SURE API is only a brick. Among other things, PIC-SURE also offers a graphical user interface that allows researchers to explore variables across multiple studies, filter patients that match criteria, and create cohorts from this interactive exploration.\n",
    "\n",
    "The API is actively developed by the Avillach-Lab at Harvard Medical School.\n",
    "\n",
    "GitHub repo:\n",
    "\n",
    "* https://github.com/hms-dbmi/pic-sure-r-adapter-hpds\n",
    "* https://github.com/hms-dbmi/pic-sure-r-client\n",
    "* https://github.com/hms-dbmi/pic-sure-biodatacatalyst-r-adapter-hpds\n",
    "\n",
    "---\n",
    "\n",
    "## Getting your own user-specific security token\n",
    "**Before running this notebook, please be sure to review the \"Get your security token\" documentation, which exists in the NHLBI_BioData_Catalyst [README.md file](https://github.com/hms-dbmi/Access-to-Data-using-PIC-SURE-API/tree/master/NHLBI_BioData_Catalyst#get-your-security-token). It explains about how to get a security token, which is mandatory to access the databases.**\n",
    "\n",
    "### Environment set-up\n",
    "\n",
    "#### Pre-requisites: \n",
    "* R version >= 3.6\n",
    "\n",
    "#### Package installation and imports\n",
    "The installation of some packages may take some time, please be patient.\n",
    "- packages listed in the `requirements.R` file\n",
    "- PIC-SURE API components (from Github)\n",
    "    - PIC-SURE Adapter\n",
    "    - PIC-SURE Client\n",
    "\n",
    "#### Install latest R PIC-SURE API libraries from GitHub\n",
    "To install the PIC-SURE libraries from GitHub, we need to install first the `devtools` package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "system(command = 'conda install -c conda-forge r-devtools --yes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Warning message in extract_lang(f = comp_lang, y = quote(if (.isMethodsDispatchOn() && :\n",
      "“Devtools is incompatible with the current version of R. `load_all()` may function incorrectly.”\n",
      "Downloading GitHub repo hms-dbmi/pic-sure-r-client@master\n",
      "from URL https://api.github.com/repos/hms-dbmi/pic-sure-r-client/zipball/master\n",
      "\n",
      "Installing picsure\n",
      "\n",
      "'/home/ec2-user/anaconda3/envs/R/lib/R/bin/R' --no-site-file --no-environ  \\\n",
      "  --no-save --no-restore --quiet CMD INSTALL  \\\n",
      "  '/tmp/Rtmp32wsAO/devtools444a291c8774/hms-dbmi-pic-sure-r-client-115deb5'  \\\n",
      "  --library='/home/ec2-user/anaconda3/envs/R/lib/R/library' --install-tests \n",
      "\n",
      "\n",
      "\n",
      "Downloading GitHub repo hms-dbmi/pic-sure-r-adapter-hpds@master\n",
      "from URL https://api.github.com/repos/hms-dbmi/pic-sure-r-adapter-hpds/zipball/master\n",
      "\n",
      "Installing hpds\n",
      "\n",
      "'/home/ec2-user/anaconda3/envs/R/lib/R/bin/R' --no-site-file --no-environ  \\\n",
      "  --no-save --no-restore --quiet CMD INSTALL  \\\n",
      "  '/tmp/Rtmp32wsAO/devtools444a6e8e77ba/hms-dbmi-pic-sure-r-adapter-hpds-2cee5ee'  \\\n",
      "  --library='/home/ec2-user/anaconda3/envs/R/lib/R/library' --install-tests \n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "Sys.setenv(TAR = \"/bin/tar\")\n",
    "options(unzip = \"internal\")\n",
    "install.packages(\"https://cran.r-project.org/src/contrib/Archive/devtools/devtools_1.13.6.tar.gz\", repos=NULL, type=\"source\")\n",
    "install.packages(\"https://cran.r-project.org/src/contrib/R6_2.5.0.tar.gz\", repos=NULL, type=\"source\")\n",
    "install.packages(\"https://cran.r-project.org/src/contrib/hash_2.2.6.1.tar.gz\", repos=NULL, type=\"source\")\n",
    "install.packages(c(\"urltools\"),repos = \"http://cran.us.r-project.org\")\n",
    "devtools::install_github(\"hms-dbmi/pic-sure-r-client\", force=T)\n",
    "devtools::install_github(\"hms-dbmi/pic-sure-r-adapter-hpds\", force=T)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Load user-defined functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# R_lib for pic-sure\n",
    "source(\"R_lib/utils.R\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Connecting to a PIC-SURE resource\n",
    "\n",
    "### 1. Connect to the UDN data network\n",
    "The following is required to get access to data through the PIC-SURE API: \n",
    "- Network URL\n",
    "- Resource id\n",
    "- User-specific security token"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Connection to the PIC-SURE API w/ key\n",
    "# network information\n",
    "PICSURE_network_URL <- \"https://udn.hms.harvard.edu/picsure\"\n",
    "resource_id <- \"c23b6814-7e5b-48d2-80d9-65511d7d2051\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# token is the individual user key given to connect to the UDN resource\n",
    "token_file <- \"token.txt\"\n",
    "my_token <- scan(token_file, what = \"character\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"c23b6814-7e5b-48d2-80d9-65511d7d2051\"\n"
     ]
    }
   ],
   "source": [
    "# get connection object\n",
    "connection <- picsure::connect(url = PICSURE_network_URL,\n",
    "                                 token = my_token)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get resource object\n",
    "resource <- hpds::get.resource(connection,\n",
    "                               resourceUUID = resource_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Two objects are created here: a `connection` and a `resource` object.\n",
    "\n",
    "Since we will only be using a single resource, **the `resource` object is actually the only one we will need to proceed with data analysis hereafter**.\n",
    "\n",
    "It is connected to the specific data source ID we specified and enables us to query and retrieve data from this database.\n",
    "\n",
    "#### Getting help with the R PIC-SURE API\n",
    "\n",
    "You can get help with PIC-SURE library functions by using the `?` operator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get function documentation\n",
    "?hpds::get.resource()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Explore the data: data structures description\n",
    "\n",
    "There are two methods to explore the data from which the user get two different data structures: a **dictionary object** to explore variables and a **query object** to explore the patient records in UDN. \n",
    "\n",
    "**Methods**:\n",
    "\n",
    "    * Search variables: find.in.dictionary() method\n",
    "    * Retrieve data: query() methods\n",
    "\n",
    "**Data structures**:\n",
    "\n",
    "    * Dictionary object structure\n",
    "    * Query object structure\n",
    "    \n",
    "\n",
    "#### Explore variables using the _dictionary_\n",
    "\n",
    "Once a connection to the desired resource has been established, we first need to get a quick idea of which variables are available in the database. To this end, we will use the `dictionary` method of the `resource` object.\n",
    "\n",
    "A dictionary object offers the possibility to retrieve information about either matching variables according to a specific term or all available variables, using the `find.in.dictionary()` method. For instance, looking for variables containing the term 'aplasia' is done this way:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a dictionary object and search for a specific term, in this example for \"aplasia\"\n",
    "lookup <- hpds::find.in.dictionary(resource, \"aplasia\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have created the dictionary object with only variables matched by the search term. To retrieve the search result from dictionary objects we have 4 different methods: `extract.count()`, `extract.keys()`, and `extract.entries()`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "$Count\n",
      "[1] 1845\n",
      "\n",
      "$Keys\n",
      "$Keys[[1]]\n",
      "[1] \"\\\\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\\\Phenotypic abnormality\\\\Abnormality of the skeletal system\\\\Abnormality of limb bone\\\\Abnormality of limb bone morphology\\\\Abnormality of digit\\\\Abnormality of toe\\\\Abnormality of the phalanges of the toes\\\\Abnormality of the phalanges of the 5th toe\\\\Aplasia/Hypoplasia of the phalanges of the 5th toe\\\\Aplasia/Hypoplasia of the middle phalanx of the 5th toe\\\\\"\n",
      "\n",
      "$Keys[[2]]\n",
      "[1] \"\\\\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\\\Phenotypic abnormality\\\\Abnormality of limbs\\\\Abnormality of limb bone\\\\Abnormality of limb bone morphology\\\\Aplasia involving bones of the extremities\\\\Aplasia/hypoplasia involving bones of the upper limbs\\\\Aplasia/hypoplasia involving bones of the hand\\\\Aplasia/Hypoplasia of fingers\\\\Aplasia/Hypoplasia of the thumb\\\\Absent thumb\\\\\"\n",
      "\n",
      "\n",
      "$Entries\n",
      "     patientCount categorical\n",
      "1               1        TRUE\n",
      "2               1        TRUE\n",
      "3               1        TRUE\n",
      "4               6        TRUE\n",
      "5               1        TRUE\n",
      "6               2        TRUE\n",
      "7               1        TRUE\n",
      "8               2        TRUE\n",
      "9               1        TRUE\n",
      "10              1        TRUE\n",
      "11              1        TRUE\n",
      "12              1        TRUE\n",
      "13              2        TRUE\n",
      "14              1        TRUE\n",
      "15              3        TRUE\n",
      "16              4        TRUE\n",
      "17              1        TRUE\n",
      "18              1        TRUE\n",
      "19              2        TRUE\n",
      "20              1        TRUE\n",
      "21              1        TRUE\n",
      "22              1        TRUE\n",
      "23              1        TRUE\n",
      "24              1        TRUE\n",
      "25              1        TRUE\n",
      "26              1        TRUE\n",
      "27              1        TRUE\n",
      "28              1        TRUE\n",
      "29              1        TRUE\n",
      "30              1        TRUE\n",
      "31              1        TRUE\n",
      "32              1        TRUE\n",
      "33              1        TRUE\n",
      "34              1        TRUE\n",
      "35              1        TRUE\n",
      "36              3        TRUE\n",
      "37              3        TRUE\n",
      "38              1        TRUE\n",
      "39              1        TRUE\n",
      "40              1        TRUE\n",
      "41              3        TRUE\n",
      "42              1        TRUE\n",
      "43              2        TRUE\n",
      "44              1        TRUE\n",
      "45              1        TRUE\n",
      "46              8        TRUE\n",
      "47              1        TRUE\n",
      "48              1        TRUE\n",
      "49              1        TRUE\n",
      "50              1        TRUE\n",
      "51              2        TRUE\n",
      "52              1        TRUE\n",
      "53              2        TRUE\n",
      "54              1        TRUE\n",
      "55              1        TRUE\n",
      "56              1        TRUE\n",
      "57              2        TRUE\n",
      "58              1        TRUE\n",
      "59              1        TRUE\n",
      "60              3        TRUE\n",
      "61              1        TRUE\n",
      "62              1        TRUE\n",
      "63              1        TRUE\n",
      "64              1        TRUE\n",
      "65              1        TRUE\n",
      "66              1        TRUE\n",
      "67              1        TRUE\n",
      "68              1        TRUE\n",
      "69              1        TRUE\n",
      "70              2        TRUE\n",
      "71              1        TRUE\n",
      "72              1        TRUE\n",
      "73              2        TRUE\n",
      "74              1        TRUE\n",
      "75              1        TRUE\n",
      "76              4        TRUE\n",
      "77              1        TRUE\n",
      "78              2        TRUE\n",
      "79              1        TRUE\n",
      "80              1        TRUE\n",
      "81              1        TRUE\n",
      "82              2        TRUE\n",
      "83              1        TRUE\n",
      "84              2        TRUE\n",
      "85              1        TRUE\n",
      "86              1        TRUE\n",
      "87              2        TRUE\n",
      "88              1        TRUE\n",
      "89              2        TRUE\n",
      "90              1        TRUE\n",
      "91              1        TRUE\n",
      "92              1        TRUE\n",
      "93              1        TRUE\n",
      "94              1        TRUE\n",
      "95              1        TRUE\n",
      "96              1        TRUE\n",
      "97             67        TRUE\n",
      "98              1        TRUE\n",
      "99             17        TRUE\n",
      "100             1        TRUE\n",
      "101             1        TRUE\n",
      "102            17        TRUE\n",
      "103             2        TRUE\n",
      "104             1        TRUE\n",
      "105             1        TRUE\n",
      "106             1        TRUE\n",
      "107             1        TRUE\n",
      "108             1        TRUE\n",
      "109             1        TRUE\n",
      "110             1        TRUE\n",
      "111             1        TRUE\n",
      "112             8        TRUE\n",
      "113             1        TRUE\n",
      "114             1        TRUE\n",
      "115             2        TRUE\n",
      "116             1        TRUE\n",
      "117             1        TRUE\n",
      "118             1        TRUE\n",
      "119             1        TRUE\n",
      "120             1        TRUE\n",
      "121             1        TRUE\n",
      "122             1        TRUE\n",
      "123             8        TRUE\n",
      "124             1        TRUE\n",
      "125             1        TRUE\n",
      "126             2        TRUE\n",
      "127             1        TRUE\n",
      "128             1        TRUE\n",
      "129             1        TRUE\n",
      "130             2        TRUE\n",
      "131             1        TRUE\n",
      "132             1        TRUE\n",
      "133             1        TRUE\n",
      "134             1        TRUE\n",
      "135             1        TRUE\n",
      "136             4        TRUE\n",
      "137             1        TRUE\n",
      "138             1        TRUE\n",
      "139             1        TRUE\n",
      "140             2        TRUE\n",
      "141             4        TRUE\n",
      "142             1        TRUE\n",
      "143             3        TRUE\n",
      "144             1        TRUE\n",
      "145             1        TRUE\n",
      "146             1        TRUE\n",
      "147             1        TRUE\n",
      "148             1        TRUE\n",
      "149             4        TRUE\n",
      "150             1        TRUE\n",
      "151             2        TRUE\n",
      "152             2        TRUE\n",
      "153             1        TRUE\n",
      "154             1        TRUE\n",
      "155             1        TRUE\n",
      "156             2        TRUE\n",
      "157             1        TRUE\n",
      "158             1        TRUE\n",
      "159             1        TRUE\n",
      "160             1        TRUE\n",
      "161             1        TRUE\n",
      "162             3        TRUE\n",
      "163             1        TRUE\n",
      "164             1        TRUE\n",
      "165             1        TRUE\n",
      "166             1        TRUE\n",
      "167            37        TRUE\n",
      "168             1        TRUE\n",
      "169             2        TRUE\n",
      "170             3        TRUE\n",
      "171             3        TRUE\n",
      "172             1        TRUE\n",
      "173             3        TRUE\n",
      "174             1        TRUE\n",
      "175             1        TRUE\n",
      "176             2        TRUE\n",
      "177             3        TRUE\n",
      "178             1        TRUE\n",
      "179             1        TRUE\n",
      "180             1        TRUE\n",
      "181             1        TRUE\n",
      "182             1        TRUE\n",
      "183             1        TRUE\n",
      "184             1        TRUE\n",
      "185             1        TRUE\n",
      "186             2        TRUE\n",
      "187             3        TRUE\n",
      "188             1        TRUE\n",
      "189             1        TRUE\n",
      "190             1        TRUE\n",
      "191             2        TRUE\n",
      "192             1        TRUE\n",
      "193             1        TRUE\n",
      "194             1        TRUE\n",
      "195             1        TRUE\n",
      "196             1        TRUE\n",
      "197             1        TRUE\n",
      "198             1        TRUE\n",
      "199             1        TRUE\n",
      "200             1        TRUE\n",
      "201             1        TRUE\n",
      "202             2        TRUE\n",
      "203             4        TRUE\n",
      "204             2        TRUE\n",
      "205             1        TRUE\n",
      "206             1        TRUE\n",
      "207             1        TRUE\n",
      "208             1        TRUE\n",
      "209             1        TRUE\n",
      "210             1        TRUE\n",
      "211             1        TRUE\n",
      "212             1        TRUE\n",
      "213             2        TRUE\n",
      "214             1        TRUE\n",
      "215             1        TRUE\n",
      "216             1        TRUE\n",
      "217             1        TRUE\n",
      "218             1        TRUE\n",
      "219             2        TRUE\n",
      "220             1        TRUE\n",
      "221             4        TRUE\n",
      "222             1        TRUE\n",
      "223             2        TRUE\n",
      "224             1        TRUE\n",
      "225             1        TRUE\n",
      "226             2        TRUE\n",
      "227             2        TRUE\n",
      "228            16        TRUE\n",
      "229             1        TRUE\n",
      "230             1        TRUE\n",
      "231             1        TRUE\n",
      "232             2        TRUE\n",
      "233             1        TRUE\n",
      "234             1        TRUE\n",
      "235             1        TRUE\n",
      "236             1        TRUE\n",
      "237             1        TRUE\n",
      "238             1        TRUE\n",
      "239             2        TRUE\n",
      "240             1        TRUE\n",
      "241             1        TRUE\n",
      "242             1        TRUE\n",
      "243             1        TRUE\n",
      "244             1        TRUE\n",
      "245             1        TRUE\n",
      "246             4        TRUE\n",
      "247             1        TRUE\n",
      "248             1        TRUE\n",
      "249             1        TRUE\n",
      "250             1        TRUE\n",
      "251            19        TRUE\n",
      "252             1        TRUE\n",
      "253             1        TRUE\n",
      "254             2        TRUE\n",
      "255             1        TRUE\n",
      "256             1        TRUE\n",
      "257             1        TRUE\n",
      "258             1        TRUE\n",
      "259             1        TRUE\n",
      "260             3        TRUE\n",
      "261             1        TRUE\n",
      "262             2        TRUE\n",
      "263             2        TRUE\n",
      "264             1        TRUE\n",
      "265             1        TRUE\n",
      "266             1        TRUE\n",
      "267             4        TRUE\n",
      "268             1        TRUE\n",
      "269             1        TRUE\n",
      "270             1        TRUE\n",
      "271             1        TRUE\n",
      "272             1        TRUE\n",
      "273             1        TRUE\n",
      "274             1        TRUE\n",
      "275             1        TRUE\n",
      "276             6        TRUE\n",
      "277             1        TRUE\n",
      "278             1        TRUE\n",
      "279             2        TRUE\n",
      "280             1        TRUE\n",
      "281             1        TRUE\n",
      "282             1        TRUE\n",
      "283             1        TRUE\n",
      "284             1        TRUE\n",
      "285             1        TRUE\n",
      "286             4        TRUE\n",
      "287             1        TRUE\n",
      "288             1        TRUE\n",
      "289             1        TRUE\n",
      "290             1        TRUE\n",
      "291             1        TRUE\n",
      "292             1        TRUE\n",
      "293             3        TRUE\n",
      "294             1        TRUE\n",
      "295             1        TRUE\n",
      "296            23        TRUE\n",
      "297             1        TRUE\n",
      "298             1        TRUE\n",
      "299             1        TRUE\n",
      "300             4        TRUE\n",
      "301             1        TRUE\n",
      "302             1        TRUE\n",
      "303             3        TRUE\n",
      "304             1        TRUE\n",
      "305             1        TRUE\n",
      "306             1        TRUE\n",
      "307             1        TRUE\n",
      "308             1        TRUE\n",
      "309             1        TRUE\n",
      "310             1        TRUE\n",
      "311             1        TRUE\n",
      "312             1        TRUE\n",
      "313             1        TRUE\n",
      "314             1        TRUE\n",
      "315             2        TRUE\n",
      "316             4        TRUE\n",
      "317             1        TRUE\n",
      "318             1        TRUE\n",
      "319             2        TRUE\n",
      "320             1        TRUE\n",
      "321             1        TRUE\n",
      "322             1        TRUE\n",
      "323             1        TRUE\n",
      "324             1        TRUE\n",
      "325             1        TRUE\n",
      "326             1        TRUE\n",
      "327             4        TRUE\n",
      "328             1        TRUE\n",
      "329             1        TRUE\n",
      "330             2        TRUE\n",
      "331             4        TRUE\n",
      "332             1        TRUE\n",
      "333             1        TRUE\n",
      "334             1        TRUE\n",
      "335             4        TRUE\n",
      "336             1        TRUE\n",
      "337             2        TRUE\n",
      "338             1        TRUE\n",
      "339             1        TRUE\n",
      "340             1        TRUE\n",
      "341             1        TRUE\n",
      "342             4        TRUE\n",
      "343             3        TRUE\n",
      "344             6        TRUE\n",
      "345             1        TRUE\n",
      "346             1        TRUE\n",
      "347             1        TRUE\n",
      "348             1        TRUE\n",
      "349            98        TRUE\n",
      "350             9        TRUE\n",
      "351             2        TRUE\n",
      "352             1        TRUE\n",
      "353             1        TRUE\n",
      "354             1        TRUE\n",
      "355             1        TRUE\n",
      "356             1        TRUE\n",
      "357             1        TRUE\n",
      "358             1        TRUE\n",
      "359             2        TRUE\n",
      "360             1        TRUE\n",
      "361             4        TRUE\n",
      "362             1        TRUE\n",
      "363             1        TRUE\n",
      "364             1        TRUE\n",
      "365             1        TRUE\n",
      "366             1        TRUE\n",
      "367             1        TRUE\n",
      "368             1        TRUE\n",
      "369             1        TRUE\n",
      "370             1        TRUE\n",
      "371             1        TRUE\n",
      "372             1        TRUE\n",
      "373             1        TRUE\n",
      "374             1        TRUE\n",
      "375             1        TRUE\n",
      "376             1        TRUE\n",
      "377             1        TRUE\n",
      "378             1        TRUE\n",
      "379             1        TRUE\n",
      "380             1        TRUE\n",
      "381             1        TRUE\n",
      "382             1        TRUE\n",
      "383             1        TRUE\n",
      "384             1        TRUE\n",
      "385             1        TRUE\n",
      "386             1        TRUE\n",
      "387             1        TRUE\n",
      "388             1        TRUE\n",
      "389             1        TRUE\n",
      "390             1        TRUE\n",
      "391             1        TRUE\n",
      "392             1        TRUE\n",
      "393             1        TRUE\n",
      "394             1        TRUE\n",
      "395             1        TRUE\n",
      "396             1        TRUE\n",
      "397             1        TRUE\n",
      "398             1        TRUE\n",
      "399             1        TRUE\n",
      "400             2        TRUE\n",
      "401            37        TRUE\n",
      "402             1        TRUE\n",
      "403             1        TRUE\n",
      "404             4        TRUE\n",
      "405             1        TRUE\n",
      "406             1        TRUE\n",
      "407             1        TRUE\n",
      "408             2        TRUE\n",
      "409             1        TRUE\n",
      "410             1        TRUE\n",
      "411             1        TRUE\n",
      "412             1        TRUE\n",
      "413             2        TRUE\n",
      "414             1        TRUE\n",
      "415             1        TRUE\n",
      "416             1        TRUE\n",
      "417             1        TRUE\n",
      "418             1        TRUE\n",
      "419             3        TRUE\n",
      "420             3        TRUE\n",
      "421             1        TRUE\n",
      "422             1        TRUE\n",
      "423             1        TRUE\n",
      "424             6        TRUE\n",
      "425             1        TRUE\n",
      "426             1        TRUE\n",
      "427             1        TRUE\n",
      "428             3        TRUE\n",
      "429             1        TRUE\n",
      "430             1        TRUE\n",
      "431            19        TRUE\n",
      "432             1        TRUE\n",
      "433             1        TRUE\n",
      "434             1        TRUE\n",
      "435             1        TRUE\n",
      "436             1        TRUE\n",
      "437             1        TRUE\n",
      "438             2        TRUE\n",
      "439             1        TRUE\n",
      "440             1        TRUE\n",
      "441             1        TRUE\n",
      "442             2        TRUE\n",
      "443             1        TRUE\n",
      "444             1        TRUE\n",
      "445             6        TRUE\n",
      "446             2        TRUE\n",
      "447             1        TRUE\n",
      "448             1        TRUE\n",
      "449             1        TRUE\n",
      "450             1        TRUE\n",
      "451             1        TRUE\n",
      "452             1        TRUE\n",
      "453             1        TRUE\n",
      "454             1        TRUE\n",
      "455             2        TRUE\n",
      "456             1        TRUE\n",
      "457             1        TRUE\n",
      "458             7        TRUE\n",
      "459             3        TRUE\n",
      "460             1        TRUE\n",
      "461             1        TRUE\n",
      "462             1        TRUE\n",
      "463             4        TRUE\n",
      "464             2        TRUE\n",
      "465             1        TRUE\n",
      "466             1        TRUE\n",
      "467             1        TRUE\n",
      "468             1        TRUE\n",
      "469             1        TRUE\n",
      "470             1        TRUE\n",
      "471             1        TRUE\n",
      "472             2        TRUE\n",
      "473             1        TRUE\n",
      "474             5        TRUE\n",
      "475             1        TRUE\n",
      "476             1        TRUE\n",
      "477            67        TRUE\n",
      "478             1        TRUE\n",
      "479             1        TRUE\n",
      "480            67        TRUE\n",
      "481             1        TRUE\n",
      "482             1        TRUE\n",
      "483             1        TRUE\n",
      "484             1        TRUE\n",
      "485             1        TRUE\n",
      "486             6        TRUE\n",
      "487             1        TRUE\n",
      "488             1        TRUE\n",
      "489             1        TRUE\n",
      "490             1        TRUE\n",
      "491             1        TRUE\n",
      "492             1        TRUE\n",
      "493             4        TRUE\n",
      "494             1        TRUE\n",
      "495             1        TRUE\n",
      "496             1        TRUE\n",
      "497             1        TRUE\n",
      "498             1        TRUE\n",
      "499             2        TRUE\n",
      "500             1        TRUE\n",
      "501             1        TRUE\n",
      "502             1        TRUE\n",
      "503             1        TRUE\n",
      "504             1        TRUE\n",
      "505             1        TRUE\n",
      "506             1        TRUE\n",
      "507             1        TRUE\n",
      "508             4        TRUE\n",
      "509             1        TRUE\n",
      "510             1        TRUE\n",
      "511             1        TRUE\n",
      "512             1        TRUE\n",
      "513             1        TRUE\n",
      "514             1        TRUE\n",
      "515             1        TRUE\n",
      "516             1        TRUE\n",
      "517             1        TRUE\n",
      "518             4        TRUE\n",
      "519             4        TRUE\n",
      "520             1        TRUE\n",
      "521             1        TRUE\n",
      "522            19        TRUE\n",
      "523            98        TRUE\n",
      "524             1        TRUE\n",
      "525             6        TRUE\n",
      "526             4        TRUE\n",
      "527             1        TRUE\n",
      "528             1        TRUE\n",
      "529             1        TRUE\n",
      "530             1        TRUE\n",
      "531            14        TRUE\n",
      "532             4        TRUE\n",
      "533             3        TRUE\n",
      "534             1        TRUE\n",
      "535             1        TRUE\n",
      "536             1        TRUE\n",
      "537             1        TRUE\n",
      "538             6        TRUE\n",
      "539             1        TRUE\n",
      "540             1        TRUE\n",
      "541             4        TRUE\n",
      "542             1        TRUE\n",
      "543             1        TRUE\n",
      "544             1        TRUE\n",
      "545             1        TRUE\n",
      "546             1        TRUE\n",
      "547             1        TRUE\n",
      "548             1        TRUE\n",
      "549             1        TRUE\n",
      "550             3        TRUE\n",
      "551             2        TRUE\n",
      "552             1        TRUE\n",
      "553             1        TRUE\n",
      "554             1        TRUE\n",
      "555             1        TRUE\n",
      "556             1        TRUE\n",
      "557             1        TRUE\n",
      "558             1        TRUE\n",
      "559             3        TRUE\n",
      "560             1        TRUE\n",
      "561             1        TRUE\n",
      "562             1        TRUE\n",
      "563             2        TRUE\n",
      "564             1        TRUE\n",
      "565             4        TRUE\n",
      "566             1        TRUE\n",
      "567             2        TRUE\n",
      "568             1        TRUE\n",
      "569             1        TRUE\n",
      "570             1        TRUE\n",
      "571             1        TRUE\n",
      "572             1        TRUE\n",
      "573            18        TRUE\n",
      "574             1        TRUE\n",
      "575             1        TRUE\n",
      "576             1        TRUE\n",
      "577             1        TRUE\n",
      "578             4        TRUE\n",
      "579             1        TRUE\n",
      "580             1        TRUE\n",
      "581             1        TRUE\n",
      "582             1        TRUE\n",
      "583             1        TRUE\n",
      "584             3        TRUE\n",
      "585             2        TRUE\n",
      "586             2        TRUE\n",
      "587             1        TRUE\n",
      "588             1        TRUE\n",
      "589             1        TRUE\n",
      "590             2        TRUE\n",
      "591             1        TRUE\n",
      "592             1        TRUE\n",
      "593             1        TRUE\n",
      "594             1        TRUE\n",
      "595             1        TRUE\n",
      "596             1        TRUE\n",
      "597             1        TRUE\n",
      "598             1        TRUE\n",
      "599            19        TRUE\n",
      "600             1        TRUE\n",
      "601             1        TRUE\n",
      "602             1        TRUE\n",
      "603             1        TRUE\n",
      "604             1        TRUE\n",
      "605             1        TRUE\n",
      "606             2        TRUE\n",
      "607             1        TRUE\n",
      "608             1        TRUE\n",
      "609             4        TRUE\n",
      "610             3        TRUE\n",
      "611            19        TRUE\n",
      "612             1        TRUE\n",
      "613             1        TRUE\n",
      "614             2        TRUE\n",
      "615             1        TRUE\n",
      "616             1        TRUE\n",
      "617             4        TRUE\n",
      "618             3        TRUE\n",
      "619             2        TRUE\n",
      "620             1        TRUE\n",
      "621             2        TRUE\n",
      "622             1        TRUE\n",
      "623             1        TRUE\n",
      "624             1        TRUE\n",
      "625             3        TRUE\n",
      "626             1        TRUE\n",
      "627             1        TRUE\n",
      "628             1        TRUE\n",
      "629             1        TRUE\n",
      "630             4        TRUE\n",
      "631             1        TRUE\n",
      "632             1        TRUE\n",
      "633             2        TRUE\n",
      "634             1        TRUE\n",
      "635             1        TRUE\n",
      "636             2        TRUE\n",
      "637             2        TRUE\n",
      "638             2        TRUE\n",
      "639             1        TRUE\n",
      "640             1        TRUE\n",
      "641             3        TRUE\n",
      "642             2        TRUE\n",
      "643             1        TRUE\n",
      "644             1        TRUE\n",
      "645             4        TRUE\n",
      "646             1        TRUE\n",
      "647             1        TRUE\n",
      "648             1        TRUE\n",
      "649             1        TRUE\n",
      "650             2        TRUE\n",
      "651             2        TRUE\n",
      "652             1        TRUE\n",
      "653             1        TRUE\n",
      "654             1        TRUE\n",
      "655             1        TRUE\n",
      "656             1        TRUE\n",
      "657             3        TRUE\n",
      "658             2        TRUE\n",
      "659             1        TRUE\n",
      "660             1        TRUE\n",
      "661             1        TRUE\n",
      "662             1        TRUE\n",
      "663            19        TRUE\n",
      "664             1        TRUE\n",
      "665             1        TRUE\n",
      "666             1        TRUE\n",
      "667             2        TRUE\n",
      "668             1        TRUE\n",
      "669             1        TRUE\n",
      "670             3        TRUE\n",
      "671             1        TRUE\n",
      "672             1        TRUE\n",
      "673             4        TRUE\n",
      "674             4        TRUE\n",
      "675             1        TRUE\n",
      "676             1        TRUE\n",
      "677             1        TRUE\n",
      "678             1        TRUE\n",
      "679             1        TRUE\n",
      "680             3        TRUE\n",
      "681             1        TRUE\n",
      "682             1        TRUE\n",
      "683             1        TRUE\n",
      "684             1        TRUE\n",
      "685             1        TRUE\n",
      "686             1        TRUE\n",
      "687             1        TRUE\n",
      "688             1        TRUE\n",
      "689             4        TRUE\n",
      "690             1        TRUE\n",
      "691             1        TRUE\n",
      "692             1        TRUE\n",
      "693             1        TRUE\n",
      "694             2        TRUE\n",
      "695             2        TRUE\n",
      "696             2        TRUE\n",
      "697             1        TRUE\n",
      "698             1        TRUE\n",
      "699             1        TRUE\n",
      "700             1        TRUE\n",
      "701             1        TRUE\n",
      "702             1        TRUE\n",
      "703             4        TRUE\n",
      "704             1        TRUE\n",
      "705             1        TRUE\n",
      "706             1        TRUE\n",
      "707             1        TRUE\n",
      "708             1        TRUE\n",
      "709             1        TRUE\n",
      "710             2        TRUE\n",
      "711             6        TRUE\n",
      "712             1        TRUE\n",
      "713            18        TRUE\n",
      "714             4        TRUE\n",
      "715             2        TRUE\n",
      "716             1        TRUE\n",
      "717             1        TRUE\n",
      "718             1        TRUE\n",
      "719             1        TRUE\n",
      "720             1        TRUE\n",
      "721            17        TRUE\n",
      "722             2        TRUE\n",
      "723             1        TRUE\n",
      "724             4        TRUE\n",
      "725             1        TRUE\n",
      "726             3        TRUE\n",
      "727             1        TRUE\n",
      "728             4        TRUE\n",
      "729             1        TRUE\n",
      "730             1        TRUE\n",
      "731            14        TRUE\n",
      "732             1        TRUE\n",
      "733             1        TRUE\n",
      "734             1        TRUE\n",
      "735             1        TRUE\n",
      "736             1        TRUE\n",
      "737             3        TRUE\n",
      "738             1        TRUE\n",
      "739             1        TRUE\n",
      "740             4        TRUE\n",
      "741             2        TRUE\n",
      "742             4        TRUE\n",
      "743             1        TRUE\n",
      "744             1        TRUE\n",
      "745             4        TRUE\n",
      "746             1        TRUE\n",
      "747             1        TRUE\n",
      "748             1        TRUE\n",
      "749             1        TRUE\n",
      "750             1        TRUE\n",
      "751             6        TRUE\n",
      "752             4        TRUE\n",
      "753             2        TRUE\n",
      "754             1        TRUE\n",
      "755             1        TRUE\n",
      "756             1        TRUE\n",
      "757             4        TRUE\n",
      "758             1        TRUE\n",
      "759             1        TRUE\n",
      "760             3        TRUE\n",
      "761            98        TRUE\n",
      "762             1        TRUE\n",
      "763             1        TRUE\n",
      "764             1        TRUE\n",
      "765             3        TRUE\n",
      "766             1        TRUE\n",
      "767             1        TRUE\n",
      "768             1        TRUE\n",
      "769             1        TRUE\n",
      "770             3        TRUE\n",
      "771             2        TRUE\n",
      "772             1        TRUE\n",
      "773             1        TRUE\n",
      "774             1        TRUE\n",
      "775             1        TRUE\n",
      "776            19        TRUE\n",
      "777             1        TRUE\n",
      "778             6        TRUE\n",
      "779             1        TRUE\n",
      "780            19        TRUE\n",
      "781             1        TRUE\n",
      "782             1        TRUE\n",
      "783             1        TRUE\n",
      "784             1        TRUE\n",
      "785             4        TRUE\n",
      "786             1        TRUE\n",
      "787             1        TRUE\n",
      "788             1        TRUE\n",
      "789             1        TRUE\n",
      "790             9        TRUE\n",
      "791             3        TRUE\n",
      "792             4        TRUE\n",
      "793             1        TRUE\n",
      "794             1        TRUE\n",
      "795             1        TRUE\n",
      "796             1        TRUE\n",
      "797             4        TRUE\n",
      "798             1        TRUE\n",
      "799             4        TRUE\n",
      "800             2        TRUE\n",
      "801             1        TRUE\n",
      "802             1        TRUE\n",
      "803             6        TRUE\n",
      "804             1        TRUE\n",
      "805            98        TRUE\n",
      "806             1        TRUE\n",
      "807             1        TRUE\n",
      "808             1        TRUE\n",
      "809             1        TRUE\n",
      "810             1        TRUE\n",
      "811             1        TRUE\n",
      "812             6        TRUE\n",
      "813             2        TRUE\n",
      "814             2        TRUE\n",
      "815             1        TRUE\n",
      "816             1        TRUE\n",
      "817             1        TRUE\n",
      "818             1        TRUE\n",
      "819             1        TRUE\n",
      "820             1        TRUE\n",
      "821             2        TRUE\n",
      "822             1        TRUE\n",
      "823             1        TRUE\n",
      "824             1        TRUE\n",
      "825             1        TRUE\n",
      "826             1        TRUE\n",
      "827             1        TRUE\n",
      "828             1        TRUE\n",
      "829             1        TRUE\n",
      "830             1        TRUE\n",
      "831             1        TRUE\n",
      "832             1        TRUE\n",
      "833             1        TRUE\n",
      "834             1        TRUE\n",
      "835             2        TRUE\n",
      "836             2        TRUE\n",
      "837             1        TRUE\n",
      "838             1        TRUE\n",
      "839             1        TRUE\n",
      "840             1        TRUE\n",
      "841             1        TRUE\n",
      "842             1        TRUE\n",
      "843             1        TRUE\n",
      "844             2        TRUE\n",
      "845             1        TRUE\n",
      "846             1        TRUE\n",
      "847             4        TRUE\n",
      "848             1        TRUE\n",
      "849             1        TRUE\n",
      "850             2        TRUE\n",
      "851             4        TRUE\n",
      "852             2        TRUE\n",
      "853             1        TRUE\n",
      "854             1        TRUE\n",
      "855             6        TRUE\n",
      "856             2        TRUE\n",
      "857             1        TRUE\n",
      "858             1        TRUE\n",
      "859            16        TRUE\n",
      "860             1        TRUE\n",
      "861             1        TRUE\n",
      "862             2        TRUE\n",
      "863             1        TRUE\n",
      "864             1        TRUE\n",
      "865             1        TRUE\n",
      "866             1        TRUE\n",
      "867             1        TRUE\n",
      "868             1        TRUE\n",
      "869             1        TRUE\n",
      "870             1        TRUE\n",
      "871             1        TRUE\n",
      "872             1        TRUE\n",
      "873             1        TRUE\n",
      "874             1        TRUE\n",
      "875             1        TRUE\n",
      "876             1        TRUE\n",
      "877             1        TRUE\n",
      "878             2        TRUE\n",
      "879             3        TRUE\n",
      "880             2        TRUE\n",
      "881             1        TRUE\n",
      "882             1        TRUE\n",
      "883             1        TRUE\n",
      "884             1        TRUE\n",
      "885             1        TRUE\n",
      "886             1        TRUE\n",
      "887             1        TRUE\n",
      "888             1        TRUE\n",
      "889             6        TRUE\n",
      "890             1        TRUE\n",
      "891             3        TRUE\n",
      "892             1        TRUE\n",
      "893             2        TRUE\n",
      "894             1        TRUE\n",
      "895             1        TRUE\n",
      "896             1        TRUE\n",
      "897             1        TRUE\n",
      "898             1        TRUE\n",
      "899             5        TRUE\n",
      "900             1        TRUE\n",
      "901             1        TRUE\n",
      "902             1        TRUE\n",
      "903             1        TRUE\n",
      "904             1        TRUE\n",
      "905             2        TRUE\n",
      "906             6        TRUE\n",
      "907             1        TRUE\n",
      "908             1        TRUE\n",
      "909             1        TRUE\n",
      "910             1        TRUE\n",
      "911             4        TRUE\n",
      "912             3        TRUE\n",
      "913             1        TRUE\n",
      "914             3        TRUE\n",
      "915             2        TRUE\n",
      "916             1        TRUE\n",
      "917             1        TRUE\n",
      "918             1        TRUE\n",
      "919             1        TRUE\n",
      "920             1        TRUE\n",
      "921             1        TRUE\n",
      "922             1        TRUE\n",
      "923             1        TRUE\n",
      "924             1        TRUE\n",
      "925             3        TRUE\n",
      "926             1        TRUE\n",
      "927             1        TRUE\n",
      "928             1        TRUE\n",
      "929            37        TRUE\n",
      "930             1        TRUE\n",
      "931             1        TRUE\n",
      "932             1        TRUE\n",
      "933             1        TRUE\n",
      "934             1        TRUE\n",
      "935             1        TRUE\n",
      "936             1        TRUE\n",
      "937             1        TRUE\n",
      "938             1        TRUE\n",
      "939             1        TRUE\n",
      "940             2        TRUE\n",
      "941             1        TRUE\n",
      "942             1        TRUE\n",
      "943             1        TRUE\n",
      "944             1        TRUE\n",
      "945            98        TRUE\n",
      "946             1        TRUE\n",
      "947             1        TRUE\n",
      "948             1        TRUE\n",
      "949             1        TRUE\n",
      "950             1        TRUE\n",
      "951             1        TRUE\n",
      "952             1        TRUE\n",
      "953             1        TRUE\n",
      "954             2        TRUE\n",
      "955             1        TRUE\n",
      "956             1        TRUE\n",
      "957             1        TRUE\n",
      "958             1        TRUE\n",
      "959             1        TRUE\n",
      "960             1        TRUE\n",
      "961             1        TRUE\n",
      "962             1        TRUE\n",
      "963             1        TRUE\n",
      "964             1        TRUE\n",
      "965             1        TRUE\n",
      "966             1        TRUE\n",
      "967             1        TRUE\n",
      "968             1        TRUE\n",
      "969             6        TRUE\n",
      "970             5        TRUE\n",
      "971             4        TRUE\n",
      "972             1        TRUE\n",
      "973             1        TRUE\n",
      "974             1        TRUE\n",
      "975             1        TRUE\n",
      "976             3        TRUE\n",
      "977             6        TRUE\n",
      "978             2        TRUE\n",
      "979             2        TRUE\n",
      "980             4        TRUE\n",
      "981             1        TRUE\n",
      "982             4        TRUE\n",
      "983             2        TRUE\n",
      "984             1        TRUE\n",
      "985             1        TRUE\n",
      "986             1        TRUE\n",
      "987             1        TRUE\n",
      "988             2        TRUE\n",
      "989             1        TRUE\n",
      "990             1        TRUE\n",
      "991             1        TRUE\n",
      "992             6        TRUE\n",
      "993             1        TRUE\n",
      "994             1        TRUE\n",
      "995             2        TRUE\n",
      "996             1        TRUE\n",
      "997             6        TRUE\n",
      "998             1        TRUE\n",
      "999             1        TRUE\n",
      "1000            1        TRUE\n",
      "1001            6        TRUE\n",
      "1002            6        TRUE\n",
      "1003            1        TRUE\n",
      "1004            4        TRUE\n",
      "1005            1        TRUE\n",
      "1006            3        TRUE\n",
      "1007            4        TRUE\n",
      "1008            2        TRUE\n",
      "1009            1        TRUE\n",
      "1010            1        TRUE\n",
      "1011            6        TRUE\n",
      "1012            2        TRUE\n",
      "1013            1        TRUE\n",
      "1014            1        TRUE\n",
      "1015            1        TRUE\n",
      "1016            1        TRUE\n",
      "1017            1        TRUE\n",
      "1018            1        TRUE\n",
      "1019            1        TRUE\n",
      "1020            1        TRUE\n",
      "1021            1        TRUE\n",
      "1022            1        TRUE\n",
      "1023            1        TRUE\n",
      "1024            1        TRUE\n",
      "1025            1        TRUE\n",
      "1026           67        TRUE\n",
      "1027            1        TRUE\n",
      "1028            1        TRUE\n",
      "1029            1        TRUE\n",
      "1030            3        TRUE\n",
      "1031            1        TRUE\n",
      "1032            2        TRUE\n",
      "1033            1        TRUE\n",
      "1034            2        TRUE\n",
      "1035            1        TRUE\n",
      "1036            1        TRUE\n",
      "1037            1        TRUE\n",
      "1038            1        TRUE\n",
      "1039            1        TRUE\n",
      "1040            2        TRUE\n",
      "1041            1        TRUE\n",
      "1042            2        TRUE\n",
      "1043           67        TRUE\n",
      "1044            3        TRUE\n",
      "1045            1        TRUE\n",
      "1046            1        TRUE\n",
      "1047            1        TRUE\n",
      "1048            1        TRUE\n",
      "1049            1        TRUE\n",
      "1050            1        TRUE\n",
      "1051            1        TRUE\n",
      "1052            2        TRUE\n",
      "1053            1        TRUE\n",
      "1054            1        TRUE\n",
      "1055            6        TRUE\n",
      "1056            1        TRUE\n",
      "1057            1        TRUE\n",
      "1058            1        TRUE\n",
      "1059            1        TRUE\n",
      "1060            1        TRUE\n",
      "1061            1        TRUE\n",
      "1062            1        TRUE\n",
      "1063            1        TRUE\n",
      "1064            1        TRUE\n",
      "1065            1        TRUE\n",
      "1066            1        TRUE\n",
      "1067            2        TRUE\n",
      "1068            1        TRUE\n",
      "1069            1        TRUE\n",
      "1070            1        TRUE\n",
      "1071            1        TRUE\n",
      "1072            2        TRUE\n",
      "1073            1        TRUE\n",
      "1074            1        TRUE\n",
      "1075            1        TRUE\n",
      "1076            1        TRUE\n",
      "1077            1        TRUE\n",
      "1078            6        TRUE\n",
      "1079            1        TRUE\n",
      "1080            1        TRUE\n",
      "1081            1        TRUE\n",
      "1082            1        TRUE\n",
      "1083            3        TRUE\n",
      "1084            1        TRUE\n",
      "1085            1        TRUE\n",
      "1086            1        TRUE\n",
      "1087            1        TRUE\n",
      "1088            1        TRUE\n",
      "1089            1        TRUE\n",
      "1090            4        TRUE\n",
      "1091            1        TRUE\n",
      "1092            2        TRUE\n",
      "1093            1        TRUE\n",
      "1094            1        TRUE\n",
      "1095            1        TRUE\n",
      "1096            1        TRUE\n",
      "1097            1        TRUE\n",
      "1098            1        TRUE\n",
      "1099            1        TRUE\n",
      "1100            1        TRUE\n",
      "1101            1        TRUE\n",
      "1102            1        TRUE\n",
      "1103            1        TRUE\n",
      "1104            1        TRUE\n",
      "1105            6        TRUE\n",
      "1106            1        TRUE\n",
      "1107            6        TRUE\n",
      "1108           98        TRUE\n",
      "1109            1        TRUE\n",
      "1110            1        TRUE\n",
      "1111            2        TRUE\n",
      "1112            1        TRUE\n",
      "1113            1        TRUE\n",
      "1114            1        TRUE\n",
      "1115            1        TRUE\n",
      "1116            1        TRUE\n",
      "1117            1        TRUE\n",
      "1118            1        TRUE\n",
      "1119            2        TRUE\n",
      "1120            1        TRUE\n",
      "1121            4        TRUE\n",
      "1122            1        TRUE\n",
      "1123            1        TRUE\n",
      "1124            4        TRUE\n",
      "1125            1        TRUE\n",
      "1126            4        TRUE\n",
      "1127            1        TRUE\n",
      "1128            1        TRUE\n",
      "1129            1        TRUE\n",
      "1130            2        TRUE\n",
      "1131            1        TRUE\n",
      "1132            6        TRUE\n",
      "1133            1        TRUE\n",
      "1134            1        TRUE\n",
      "1135            4        TRUE\n",
      "1136            1        TRUE\n",
      "1137            1        TRUE\n",
      "1138            1        TRUE\n",
      "1139            1        TRUE\n",
      "1140            1        TRUE\n",
      "1141            3        TRUE\n",
      "1142            1        TRUE\n",
      "1143            1        TRUE\n",
      "1144            1        TRUE\n",
      "1145            1        TRUE\n",
      "1146            1        TRUE\n",
      "1147            2        TRUE\n",
      "1148            2        TRUE\n",
      "1149            1        TRUE\n",
      "1150           67        TRUE\n",
      "1151            9        TRUE\n",
      "1152            1        TRUE\n",
      "1153            2        TRUE\n",
      "1154            1        TRUE\n",
      "1155            2        TRUE\n",
      "1156            1        TRUE\n",
      "1157            1        TRUE\n",
      "1158            1        TRUE\n",
      "1159            1        TRUE\n",
      "1160            1        TRUE\n",
      "1161            2        TRUE\n",
      "1162            1        TRUE\n",
      "1163            2        TRUE\n",
      "1164           17        TRUE\n",
      "1165            4        TRUE\n",
      "1166            1        TRUE\n",
      "1167            1        TRUE\n",
      "1168            3        TRUE\n",
      "1169            1        TRUE\n",
      "1170            1        TRUE\n",
      "1171            1        TRUE\n",
      "1172            1        TRUE\n",
      "1173            1        TRUE\n",
      "1174            1        TRUE\n",
      "1175            1        TRUE\n",
      "1176            1        TRUE\n",
      "1177            1        TRUE\n",
      "1178            1        TRUE\n",
      "1179            1        TRUE\n",
      "1180            1        TRUE\n",
      "1181            1        TRUE\n",
      "1182            1        TRUE\n",
      "1183            1        TRUE\n",
      "1184            1        TRUE\n",
      "1185            1        TRUE\n",
      "1186            1        TRUE\n",
      "1187            2        TRUE\n",
      "1188            3        TRUE\n",
      "1189            1        TRUE\n",
      "1190            1        TRUE\n",
      "1191            2        TRUE\n",
      "1192            1        TRUE\n",
      "1193            4        TRUE\n",
      "1194            1        TRUE\n",
      "1195            1        TRUE\n",
      "1196            1        TRUE\n",
      "1197            1        TRUE\n",
      "1198            1        TRUE\n",
      "1199            1        TRUE\n",
      "1200            1        TRUE\n",
      "1201            1        TRUE\n",
      "1202            1        TRUE\n",
      "1203            1        TRUE\n",
      "1204            1        TRUE\n",
      "1205            2        TRUE\n",
      "1206            4        TRUE\n",
      "1207           37        TRUE\n",
      "1208            1        TRUE\n",
      "1209            1        TRUE\n",
      "1210            1        TRUE\n",
      "1211            1        TRUE\n",
      "1212            2        TRUE\n",
      "1213            3        TRUE\n",
      "1214            1        TRUE\n",
      "1215            8        TRUE\n",
      "1216            1        TRUE\n",
      "1217            1        TRUE\n",
      "1218            1        TRUE\n",
      "1219            1        TRUE\n",
      "1220            1        TRUE\n",
      "1221            1        TRUE\n",
      "1222            1        TRUE\n",
      "1223            1        TRUE\n",
      "1224            1        TRUE\n",
      "1225            1        TRUE\n",
      "1226            1        TRUE\n",
      "1227            5        TRUE\n",
      "1228            1        TRUE\n",
      "1229            2        TRUE\n",
      "1230           19        TRUE\n",
      "1231            3        TRUE\n",
      "1232            1        TRUE\n",
      "1233            1        TRUE\n",
      "1234            1        TRUE\n",
      "1235            1        TRUE\n",
      "1236            1        TRUE\n",
      "1237            1        TRUE\n",
      "1238            4        TRUE\n",
      "1239            1        TRUE\n",
      "1240            1        TRUE\n",
      "1241            1        TRUE\n",
      "1242            1        TRUE\n",
      "1243            1        TRUE\n",
      "1244            1        TRUE\n",
      "1245            1        TRUE\n",
      "1246            1        TRUE\n",
      "1247            1        TRUE\n",
      "1248            1        TRUE\n",
      "1249            1        TRUE\n",
      "1250            4        TRUE\n",
      "1251            1        TRUE\n",
      "1252            1        TRUE\n",
      "1253            1        TRUE\n",
      "1254            3        TRUE\n",
      "1255            1        TRUE\n",
      "1256            4        TRUE\n",
      "1257            2        TRUE\n",
      "1258            1        TRUE\n",
      "1259            1        TRUE\n",
      "1260            1        TRUE\n",
      "1261            1        TRUE\n",
      "1262            2        TRUE\n",
      "1263            1        TRUE\n",
      "1264            1        TRUE\n",
      "1265            1        TRUE\n",
      "1266            1        TRUE\n",
      "1267            3        TRUE\n",
      "1268            2        TRUE\n",
      "1269            1        TRUE\n",
      "1270            1        TRUE\n",
      "1271            1        TRUE\n",
      "1272            2        TRUE\n",
      "1273            1        TRUE\n",
      "1274            1        TRUE\n",
      "1275            1        TRUE\n",
      "1276            1        TRUE\n",
      "1277            1        TRUE\n",
      "1278            1        TRUE\n",
      "1279            1        TRUE\n",
      "1280            1        TRUE\n",
      "1281            1        TRUE\n",
      "1282            2        TRUE\n",
      "1283            1        TRUE\n",
      "1284            1        TRUE\n",
      "1285           37        TRUE\n",
      "1286            1        TRUE\n",
      "1287            1        TRUE\n",
      "1288            2        TRUE\n",
      "1289            1        TRUE\n",
      "1290            2        TRUE\n",
      "1291            1        TRUE\n",
      "1292            2        TRUE\n",
      "1293            3        TRUE\n",
      "1294            1        TRUE\n",
      "1295            2        TRUE\n",
      "1296            1        TRUE\n",
      "1297            1        TRUE\n",
      "1298            1        TRUE\n",
      "1299            1        TRUE\n",
      "1300            1        TRUE\n",
      "1301            1        TRUE\n",
      "1302            1        TRUE\n",
      "1303            1        TRUE\n",
      "1304            1        TRUE\n",
      "1305            1        TRUE\n",
      "1306            1        TRUE\n",
      "1307            1        TRUE\n",
      "1308            1        TRUE\n",
      "1309            4        TRUE\n",
      "1310            1        TRUE\n",
      "1311            1        TRUE\n",
      "1312            1        TRUE\n",
      "1313            1        TRUE\n",
      "1314            2        TRUE\n",
      "1315            1        TRUE\n",
      "1316            1        TRUE\n",
      "1317            1        TRUE\n",
      "1318            1        TRUE\n",
      "1319            1        TRUE\n",
      "1320            2        TRUE\n",
      "1321            1        TRUE\n",
      "1322            1        TRUE\n",
      "1323            1        TRUE\n",
      "1324            6        TRUE\n",
      "1325            4        TRUE\n",
      "1326            1        TRUE\n",
      "1327            1        TRUE\n",
      "1328           19        TRUE\n",
      "1329            4        TRUE\n",
      "1330            1        TRUE\n",
      "1331            2        TRUE\n",
      "1332            1        TRUE\n",
      "1333            1        TRUE\n",
      "1334            1        TRUE\n",
      "1335            1        TRUE\n",
      "1336            1        TRUE\n",
      "1337            1        TRUE\n",
      "1338            4        TRUE\n",
      "1339            2        TRUE\n",
      "1340            1        TRUE\n",
      "1341            1        TRUE\n",
      "1342            1        TRUE\n",
      "1343            1        TRUE\n",
      "1344            1        TRUE\n",
      "1345            1        TRUE\n",
      "1346            1        TRUE\n",
      "1347            1        TRUE\n",
      "1348            3        TRUE\n",
      "1349            2        TRUE\n",
      "1350            6        TRUE\n",
      "1351            9        TRUE\n",
      "1352            1        TRUE\n",
      "1353            1        TRUE\n",
      "1354            3        TRUE\n",
      "1355            6        TRUE\n",
      "1356            1        TRUE\n",
      "1357            1        TRUE\n",
      "1358            1        TRUE\n",
      "1359            1        TRUE\n",
      "1360            1        TRUE\n",
      "1361            1        TRUE\n",
      "1362            3        TRUE\n",
      "1363            1        TRUE\n",
      "1364            1        TRUE\n",
      "1365            2        TRUE\n",
      "1366            1        TRUE\n",
      "1367            1        TRUE\n",
      "1368            1        TRUE\n",
      "1369            1        TRUE\n",
      "1370            1        TRUE\n",
      "1371            1        TRUE\n",
      "1372            1        TRUE\n",
      "1373            1        TRUE\n",
      "1374            1        TRUE\n",
      "1375            1        TRUE\n",
      "1376            1        TRUE\n",
      "1377            2        TRUE\n",
      "1378            2        TRUE\n",
      "1379            4        TRUE\n",
      "1380            1        TRUE\n",
      "1381            1        TRUE\n",
      "1382            1        TRUE\n",
      "1383            1        TRUE\n",
      "1384            1        TRUE\n",
      "1385            1        TRUE\n",
      "1386            3        TRUE\n",
      "1387            1        TRUE\n",
      "1388            1        TRUE\n",
      "1389            4        TRUE\n",
      "1390            4        TRUE\n",
      "1391            1        TRUE\n",
      "1392            1        TRUE\n",
      "1393            1        TRUE\n",
      "1394            1        TRUE\n",
      "1395            3        TRUE\n",
      "1396            1        TRUE\n",
      "1397            1        TRUE\n",
      "1398            1        TRUE\n",
      "1399            1        TRUE\n",
      "1400            1        TRUE\n",
      "1401            1        TRUE\n",
      "1402            1        TRUE\n",
      "1403            1        TRUE\n",
      "1404            1        TRUE\n",
      "1405            1        TRUE\n",
      "1406            1        TRUE\n",
      "1407            3        TRUE\n",
      "1408           37        TRUE\n",
      "1409            6        TRUE\n",
      "1410            1        TRUE\n",
      "1411            4        TRUE\n",
      "1412            1        TRUE\n",
      "1413            1        TRUE\n",
      "1414            1        TRUE\n",
      "1415            1        TRUE\n",
      "1416            3        TRUE\n",
      "1417            1        TRUE\n",
      "1418            1        TRUE\n",
      "1419            2        TRUE\n",
      "1420            1        TRUE\n",
      "1421            4        TRUE\n",
      "1422            1        TRUE\n",
      "1423            1        TRUE\n",
      "1424            2        TRUE\n",
      "1425            1        TRUE\n",
      "1426            2        TRUE\n",
      "1427            1        TRUE\n",
      "1428            1        TRUE\n",
      "1429            1        TRUE\n",
      "1430            1        TRUE\n",
      "1431            9        TRUE\n",
      "1432            1        TRUE\n",
      "1433            1        TRUE\n",
      "1434            1        TRUE\n",
      "1435            1        TRUE\n",
      "1436            1        TRUE\n",
      "1437            1        TRUE\n",
      "1438            1        TRUE\n",
      "1439            1        TRUE\n",
      "1440            1        TRUE\n",
      "1441            1        TRUE\n",
      "1442            1        TRUE\n",
      "1443            6        TRUE\n",
      "1444            1        TRUE\n",
      "1445            1        TRUE\n",
      "1446            1        TRUE\n",
      "1447            3        TRUE\n",
      "1448            1        TRUE\n",
      "1449            1        TRUE\n",
      "1450            2        TRUE\n",
      "1451            4        TRUE\n",
      "1452            1        TRUE\n",
      "1453            4        TRUE\n",
      "1454            1        TRUE\n",
      "1455            1        TRUE\n",
      "1456            1        TRUE\n",
      "1457            1        TRUE\n",
      "1458            1        TRUE\n",
      "1459            1        TRUE\n",
      "1460            1        TRUE\n",
      "1461            4        TRUE\n",
      "1462            1        TRUE\n",
      "1463            1        TRUE\n",
      "1464            1        TRUE\n",
      "1465            2        TRUE\n",
      "1466            1        TRUE\n",
      "1467            1        TRUE\n",
      "1468            1        TRUE\n",
      "1469            3        TRUE\n",
      "1470            1        TRUE\n",
      "1471            1        TRUE\n",
      "1472            1        TRUE\n",
      "1473           98        TRUE\n",
      "1474            3        TRUE\n",
      "1475            4        TRUE\n",
      "1476           17        TRUE\n",
      "1477            1        TRUE\n",
      "1478            1        TRUE\n",
      "1479           37        TRUE\n",
      "1480            1        TRUE\n",
      "1481            4        TRUE\n",
      "1482            4        TRUE\n",
      "1483            1        TRUE\n",
      "1484            1        TRUE\n",
      "1485            1        TRUE\n",
      "1486            1        TRUE\n",
      "1487            1        TRUE\n",
      "1488            1        TRUE\n",
      "1489            1        TRUE\n",
      "1490            3        TRUE\n",
      "1491            1        TRUE\n",
      "1492            1        TRUE\n",
      "1493            2        TRUE\n",
      "1494            1        TRUE\n",
      "1495            1        TRUE\n",
      "1496            3        TRUE\n",
      "1497            1        TRUE\n",
      "1498            1        TRUE\n",
      "1499            1        TRUE\n",
      "1500            2        TRUE\n",
      "1501            2        TRUE\n",
      "1502            1        TRUE\n",
      "1503           19        TRUE\n",
      "1504            1        TRUE\n",
      "1505            1        TRUE\n",
      "1506            9        TRUE\n",
      "1507            1        TRUE\n",
      "1508            3        TRUE\n",
      "1509            3        TRUE\n",
      "1510           19        TRUE\n",
      "1511            1        TRUE\n",
      "1512            1        TRUE\n",
      "1513            1        TRUE\n",
      "1514            6        TRUE\n",
      "1515            1        TRUE\n",
      "1516            3        TRUE\n",
      "1517            1        TRUE\n",
      "1518            3        TRUE\n",
      "1519            1        TRUE\n",
      "1520            1        TRUE\n",
      "1521            1        TRUE\n",
      "1522            1        TRUE\n",
      "1523            2        TRUE\n",
      "1524            1        TRUE\n",
      "1525            1        TRUE\n",
      "1526            1        TRUE\n",
      "1527            1        TRUE\n",
      "1528            1        TRUE\n",
      "1529            1        TRUE\n",
      "1530            6        TRUE\n",
      "1531            2        TRUE\n",
      "1532            4        TRUE\n",
      "1533            1        TRUE\n",
      "1534            1        TRUE\n",
      "1535            1        TRUE\n",
      "1536           19        TRUE\n",
      "1537            1        TRUE\n",
      "1538            6        TRUE\n",
      "1539            1        TRUE\n",
      "1540            3        TRUE\n",
      "1541            1        TRUE\n",
      "1542            1        TRUE\n",
      "1543            1        TRUE\n",
      "1544            2        TRUE\n",
      "1545            1        TRUE\n",
      "1546            2        TRUE\n",
      "1547            1        TRUE\n",
      "1548            1        TRUE\n",
      "1549            1        TRUE\n",
      "1550            1        TRUE\n",
      "1551           20        TRUE\n",
      "1552            1        TRUE\n",
      "1553            1        TRUE\n",
      "1554            1        TRUE\n",
      "1555            1        TRUE\n",
      "1556            1        TRUE\n",
      "1557            1        TRUE\n",
      "1558            4        TRUE\n",
      "1559            1        TRUE\n",
      "1560            1        TRUE\n",
      "1561            1        TRUE\n",
      "1562            2        TRUE\n",
      "1563            3        TRUE\n",
      "1564            1        TRUE\n",
      "1565            1        TRUE\n",
      "1566            1        TRUE\n",
      "1567            1        TRUE\n",
      "1568            2        TRUE\n",
      "1569            1        TRUE\n",
      "1570            1        TRUE\n",
      "1571            1        TRUE\n",
      "1572            1        TRUE\n",
      "1573            1        TRUE\n",
      "1574            1        TRUE\n",
      "1575            1        TRUE\n",
      "1576            1        TRUE\n",
      "1577            1        TRUE\n",
      "1578            1        TRUE\n",
      "1579            1        TRUE\n",
      "1580            1        TRUE\n",
      "1581            1        TRUE\n",
      "1582            1        TRUE\n",
      "1583            1        TRUE\n",
      "1584            1        TRUE\n",
      "1585            1        TRUE\n",
      "1586            1        TRUE\n",
      "1587            3        TRUE\n",
      "1588            1        TRUE\n",
      "1589            1        TRUE\n",
      "1590            6        TRUE\n",
      "1591            1        TRUE\n",
      "1592            1        TRUE\n",
      "1593            1        TRUE\n",
      "1594            2        TRUE\n",
      "1595            1        TRUE\n",
      "1596            6        TRUE\n",
      "1597            6        TRUE\n",
      "1598            1        TRUE\n",
      "1599            1        TRUE\n",
      "1600            1        TRUE\n",
      "1601            1        TRUE\n",
      "1602            1        TRUE\n",
      "1603            1        TRUE\n",
      "1604            1        TRUE\n",
      "1605            1        TRUE\n",
      "1606            2        TRUE\n",
      "1607            1        TRUE\n",
      "1608            2        TRUE\n",
      "1609            2        TRUE\n",
      "1610            4        TRUE\n",
      "1611            2        TRUE\n",
      "1612            1        TRUE\n",
      "1613            1        TRUE\n",
      "1614            1        TRUE\n",
      "1615            1        TRUE\n",
      "1616            4        TRUE\n",
      "1617            2        TRUE\n",
      "1618            2        TRUE\n",
      "1619            1        TRUE\n",
      "1620            2        TRUE\n",
      "1621            1        TRUE\n",
      "1622            1        TRUE\n",
      "1623            2        TRUE\n",
      "1624            2        TRUE\n",
      "1625           98        TRUE\n",
      "1626            1        TRUE\n",
      "1627            1        TRUE\n",
      "1628            1        TRUE\n",
      "1629            1        TRUE\n",
      "1630            1        TRUE\n",
      "1631            1        TRUE\n",
      "1632            1        TRUE\n",
      "1633            1        TRUE\n",
      "1634            1        TRUE\n",
      "1635            1        TRUE\n",
      "1636            1        TRUE\n",
      "1637            1        TRUE\n",
      "1638            4        TRUE\n",
      "1639           67        TRUE\n",
      "1640            6        TRUE\n",
      "1641            1        TRUE\n",
      "1642            6        TRUE\n",
      "1643            4        TRUE\n",
      "1644            1        TRUE\n",
      "1645            1        TRUE\n",
      "1646            1        TRUE\n",
      "1647            3        TRUE\n",
      "1648            2        TRUE\n",
      "1649            1        TRUE\n",
      "1650            1        TRUE\n",
      "1651            1        TRUE\n",
      "1652            1        TRUE\n",
      "1653            3        TRUE\n",
      "1654            1        TRUE\n",
      "1655            1        TRUE\n",
      "1656            1        TRUE\n",
      "1657            1        TRUE\n",
      "1658            2        TRUE\n",
      "1659            1        TRUE\n",
      "1660            2        TRUE\n",
      "1661            1        TRUE\n",
      "1662            1        TRUE\n",
      "1663            1        TRUE\n",
      "1664            1        TRUE\n",
      "1665            1        TRUE\n",
      "1666            1        TRUE\n",
      "1667            4        TRUE\n",
      "1668            1        TRUE\n",
      "1669            1        TRUE\n",
      "1670            1        TRUE\n",
      "1671            2        TRUE\n",
      "1672            1        TRUE\n",
      "1673            1        TRUE\n",
      "1674            1        TRUE\n",
      "1675            1        TRUE\n",
      "1676            2        TRUE\n",
      "1677            1        TRUE\n",
      "1678            1        TRUE\n",
      "1679            1        TRUE\n",
      "1680            1        TRUE\n",
      "1681            1        TRUE\n",
      "1682            1        TRUE\n",
      "1683            1        TRUE\n",
      "1684            2        TRUE\n",
      "1685            2        TRUE\n",
      "1686            1        TRUE\n",
      "1687            1        TRUE\n",
      "1688            2        TRUE\n",
      "1689            2        TRUE\n",
      "1690            3        TRUE\n",
      "1691            1        TRUE\n",
      "1692            1        TRUE\n",
      "1693            2        TRUE\n",
      "1694            1        TRUE\n",
      "1695            1        TRUE\n",
      "1696            1        TRUE\n",
      "1697            1        TRUE\n",
      "1698           98        TRUE\n",
      "1699            1        TRUE\n",
      "1700            4        TRUE\n",
      "1701            2        TRUE\n",
      "1702            2        TRUE\n",
      "1703            1        TRUE\n",
      "1704            3        TRUE\n",
      "1705            1        TRUE\n",
      "1706            3        TRUE\n",
      "1707            1        TRUE\n",
      "1708            3        TRUE\n",
      "1709            1        TRUE\n",
      "1710            1        TRUE\n",
      "1711            1        TRUE\n",
      "1712            4        TRUE\n",
      "1713            2        TRUE\n",
      "1714            4        TRUE\n",
      "1715            1        TRUE\n",
      "1716            1        TRUE\n",
      "1717            1        TRUE\n",
      "1718            1        TRUE\n",
      "1719            1        TRUE\n",
      "1720            1        TRUE\n",
      "1721            1        TRUE\n",
      "1722            1        TRUE\n",
      "1723            1        TRUE\n",
      "1724            1        TRUE\n",
      "1725            4        TRUE\n",
      "1726            2        TRUE\n",
      "1727            1        TRUE\n",
      "1728            1        TRUE\n",
      "1729            1        TRUE\n",
      "1730            1        TRUE\n",
      "1731            2        TRUE\n",
      "1732            1        TRUE\n",
      "1733            1        TRUE\n",
      "1734            3        TRUE\n",
      "1735            1        TRUE\n",
      "1736            1        TRUE\n",
      "1737            1        TRUE\n",
      "1738            3        TRUE\n",
      "1739            4        TRUE\n",
      "1740            1        TRUE\n",
      "1741            1        TRUE\n",
      "1742            6        TRUE\n",
      "1743            1        TRUE\n",
      "1744            1        TRUE\n",
      "1745            2        TRUE\n",
      "1746            1        TRUE\n",
      "1747            1        TRUE\n",
      "1748            5        TRUE\n",
      "1749            1        TRUE\n",
      "1750            1        TRUE\n",
      "1751            1        TRUE\n",
      "1752            1        TRUE\n",
      "1753            1        TRUE\n",
      "1754            1        TRUE\n",
      "1755            1        TRUE\n",
      "1756            3        TRUE\n",
      "1757            1        TRUE\n",
      "1758            1        TRUE\n",
      "1759            1        TRUE\n",
      "1760            1        TRUE\n",
      "1761            3        TRUE\n",
      "1762            6        TRUE\n",
      "1763            1        TRUE\n",
      "1764            1        TRUE\n",
      "1765            2        TRUE\n",
      "1766            1        TRUE\n",
      "1767            1        TRUE\n",
      "1768            4        TRUE\n",
      "1769            1        TRUE\n",
      "1770            1        TRUE\n",
      "1771            1        TRUE\n",
      "1772            4        TRUE\n",
      "1773            1        TRUE\n",
      "1774            2        TRUE\n",
      "1775            1        TRUE\n",
      "1776            2        TRUE\n",
      "1777            1        TRUE\n",
      "1778            1        TRUE\n",
      "1779            2        TRUE\n",
      "1780            1        TRUE\n",
      "1781            1        TRUE\n",
      "1782            3        TRUE\n",
      "1783            5        TRUE\n",
      "1784            3        TRUE\n",
      "1785            1        TRUE\n",
      "1786            1        TRUE\n",
      "1787            1        TRUE\n",
      "1788            1        TRUE\n",
      "1789            1        TRUE\n",
      "1790            1        TRUE\n",
      "1791            1        TRUE\n",
      "1792            1        TRUE\n",
      "1793            1        TRUE\n",
      "1794            1        TRUE\n",
      "1795            1        TRUE\n",
      "1796            1        TRUE\n",
      "1797            9        TRUE\n",
      "1798            1        TRUE\n",
      "1799            2        TRUE\n",
      "1800            1        TRUE\n",
      "1801            1        TRUE\n",
      "1802            1        TRUE\n",
      "1803            2        TRUE\n",
      "1804            1        TRUE\n",
      "1805            1        TRUE\n",
      "1806            1        TRUE\n",
      "1807            1        TRUE\n",
      "1808            3        TRUE\n",
      "1809            3        TRUE\n",
      "1810            2        TRUE\n",
      "1811            1        TRUE\n",
      "1812            1        TRUE\n",
      "1813            2        TRUE\n",
      "1814            1        TRUE\n",
      "1815            1        TRUE\n",
      "1816            1        TRUE\n",
      "1817            2        TRUE\n",
      "1818            1        TRUE\n",
      "1819            5        TRUE\n",
      "1820            1        TRUE\n",
      "1821            1        TRUE\n",
      "1822            1        TRUE\n",
      "1823            1        TRUE\n",
      "1824            1        TRUE\n",
      "1825            2        TRUE\n",
      "1826            1        TRUE\n",
      "1827            1        TRUE\n",
      "1828            1        TRUE\n",
      "1829            1        TRUE\n",
      "1830            1        TRUE\n",
      "1831            2        TRUE\n",
      "1832            1        TRUE\n",
      "1833            1        TRUE\n",
      "1834            3        TRUE\n",
      "1835            2        TRUE\n",
      "1836            6        TRUE\n",
      "1837           10        TRUE\n",
      "1838            2        TRUE\n",
      "1839            1        TRUE\n",
      "1840            1        TRUE\n",
      "1841            1        TRUE\n",
      "1842            1        TRUE\n",
      "1843            1        TRUE\n",
      "1844            1        TRUE\n",
      "1845            1        TRUE\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# description of the dictionary search content\n",
    "print(list(\"Count\"   = hpds::extract.count(lookup), \n",
    "           \"Keys\"    = hpds::extract.keys(lookup)[0:2],\n",
    "           \"Entries\" = hpds::extract.entries(lookup)[0:2]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**hpds::extract.entries()** enables us to get the result of the dictionary search in a data.frame format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>patientCount</th><th scope=col>categorical</th><th scope=col>observationCount</th><th scope=col>name</th><th scope=col>min</th><th scope=col>max</th><th scope=col>HpdsDataType</th><th scope=col>categoryValues</th><th scope=col>description</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;list&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1844</th><td>1</td><td>TRUE</td><td>1</td><td>\\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\Phenotypic abnormality\\Abnormality of the skeletal system\\Abnormality of skeletal morphology\\Aplasia/hypoplasia involving the skeleton\\Aplasia/hypoplasia of the extremities\\Aplasia involving bones of the extremities\\Aplasia/hypoplasia involving bones of the lower limbs\\Aplasia/Hypoplasia involving bones of the feet\\Aplasia/Hypoplasia of toe\\Absent toe\\Aplasia/Hypoplasia of the distal phalanges of the toes\\Aplasia/Hypoplasia of the distal phalanx of the 4th toe\\Short distal phalanx of the 4th toe\\</td><td>NA</td><td>NA</td><td>phenotypes</td><td>Positive</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>1845</th><td>1</td><td>TRUE</td><td>1</td><td>\\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\Phenotypic abnormality\\Abnormality of the skeletal system\\Abnormality of skeletal morphology\\Aplasia/hypoplasia involving the skeleton\\Aplasia/hypoplasia of the extremities\\Aplasia involving bones of the extremities\\Aplasia/hypoplasia involving bones of the lower limbs\\Aplasia/Hypoplasia involving bones of the feet\\Aplasia/Hypoplasia of toe\\Absent toe\\Aplasia/Hypoplasia of the 2nd toe\\Short phalanx of the 2nd toe\\Short distal phalanx of the 2nd toe\\                                                </td><td>NA</td><td>NA</td><td>phenotypes</td><td>Positive</td><td>NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 9\n",
       "\\begin{tabular}{r|lllllllll}\n",
       "  & patientCount & categorical & observationCount & name & min & max & HpdsDataType & categoryValues & description\\\\\n",
       "  & <int> & <lgl> & <int> & <chr> & <lgl> & <lgl> & <chr> & <list> & <lgl>\\\\\n",
       "\\hline\n",
       "\t1844 & 1 & TRUE & 1 & \\textbackslash{}04\\_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\textbackslash{}Phenotypic abnormality\\textbackslash{}Abnormality of the skeletal system\\textbackslash{}Abnormality of skeletal morphology\\textbackslash{}Aplasia/hypoplasia involving the skeleton\\textbackslash{}Aplasia/hypoplasia of the extremities\\textbackslash{}Aplasia involving bones of the extremities\\textbackslash{}Aplasia/hypoplasia involving bones of the lower limbs\\textbackslash{}Aplasia/Hypoplasia involving bones of the feet\\textbackslash{}Aplasia/Hypoplasia of toe\\textbackslash{}Absent toe\\textbackslash{}Aplasia/Hypoplasia of the distal phalanges of the toes\\textbackslash{}Aplasia/Hypoplasia of the distal phalanx of the 4th toe\\textbackslash{}Short distal phalanx of the 4th toe\\textbackslash{} & NA & NA & phenotypes & Positive & NA\\\\\n",
       "\t1845 & 1 & TRUE & 1 & \\textbackslash{}04\\_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\textbackslash{}Phenotypic abnormality\\textbackslash{}Abnormality of the skeletal system\\textbackslash{}Abnormality of skeletal morphology\\textbackslash{}Aplasia/hypoplasia involving the skeleton\\textbackslash{}Aplasia/hypoplasia of the extremities\\textbackslash{}Aplasia involving bones of the extremities\\textbackslash{}Aplasia/hypoplasia involving bones of the lower limbs\\textbackslash{}Aplasia/Hypoplasia involving bones of the feet\\textbackslash{}Aplasia/Hypoplasia of toe\\textbackslash{}Absent toe\\textbackslash{}Aplasia/Hypoplasia of the 2nd toe\\textbackslash{}Short phalanx of the 2nd toe\\textbackslash{}Short distal phalanx of the 2nd toe\\textbackslash{}                                                 & NA & NA & phenotypes & Positive & NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 9\n",
       "\n",
       "| <!--/--> | patientCount &lt;int&gt; | categorical &lt;lgl&gt; | observationCount &lt;int&gt; | name &lt;chr&gt; | min &lt;lgl&gt; | max &lt;lgl&gt; | HpdsDataType &lt;chr&gt; | categoryValues &lt;list&gt; | description &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1844 | 1 | TRUE | 1 | \\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\Phenotypic abnormality\\Abnormality of the skeletal system\\Abnormality of skeletal morphology\\Aplasia/hypoplasia involving the skeleton\\Aplasia/hypoplasia of the extremities\\Aplasia involving bones of the extremities\\Aplasia/hypoplasia involving bones of the lower limbs\\Aplasia/Hypoplasia involving bones of the feet\\Aplasia/Hypoplasia of toe\\Absent toe\\Aplasia/Hypoplasia of the distal phalanges of the toes\\Aplasia/Hypoplasia of the distal phalanx of the 4th toe\\Short distal phalanx of the 4th toe\\ | NA | NA | phenotypes | Positive | NA |\n",
       "| 1845 | 1 | TRUE | 1 | \\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\Phenotypic abnormality\\Abnormality of the skeletal system\\Abnormality of skeletal morphology\\Aplasia/hypoplasia involving the skeleton\\Aplasia/hypoplasia of the extremities\\Aplasia involving bones of the extremities\\Aplasia/hypoplasia involving bones of the lower limbs\\Aplasia/Hypoplasia involving bones of the feet\\Aplasia/Hypoplasia of toe\\Absent toe\\Aplasia/Hypoplasia of the 2nd toe\\Short phalanx of the 2nd toe\\Short distal phalanx of the 2nd toe\\                                                 | NA | NA | phenotypes | Positive | NA |\n",
       "\n"
      ],
      "text/plain": [
       "     patientCount categorical observationCount\n",
       "1844 1            TRUE        1               \n",
       "1845 1            TRUE        1               \n",
       "     name                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     \n",
       "1844 \\\\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\\\Phenotypic abnormality\\\\Abnormality of the skeletal system\\\\Abnormality of skeletal morphology\\\\Aplasia/hypoplasia involving the skeleton\\\\Aplasia/hypoplasia of the extremities\\\\Aplasia involving bones of the extremities\\\\Aplasia/hypoplasia involving bones of the lower limbs\\\\Aplasia/Hypoplasia involving bones of the feet\\\\Aplasia/Hypoplasia of toe\\\\Absent toe\\\\Aplasia/Hypoplasia of the distal phalanges of the toes\\\\Aplasia/Hypoplasia of the distal phalanx of the 4th toe\\\\Short distal phalanx of the 4th toe\\\\\n",
       "1845 \\\\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\\\Phenotypic abnormality\\\\Abnormality of the skeletal system\\\\Abnormality of skeletal morphology\\\\Aplasia/hypoplasia involving the skeleton\\\\Aplasia/hypoplasia of the extremities\\\\Aplasia involving bones of the extremities\\\\Aplasia/hypoplasia involving bones of the lower limbs\\\\Aplasia/Hypoplasia involving bones of the feet\\\\Aplasia/Hypoplasia of toe\\\\Absent toe\\\\Aplasia/Hypoplasia of the 2nd toe\\\\Short phalanx of the 2nd toe\\\\Short distal phalanx of the 2nd toe\\\\                                                \n",
       "     min max HpdsDataType categoryValues description\n",
       "1844 NA  NA  phenotypes   Positive       NA         \n",
       "1845 NA  NA  phenotypes   Positive       NA         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# show table of records from the dictionary object\n",
    "hpds::extract.entries(lookup) %>% tail(, n =2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can retrieve information about **ALL** variables. We do it without specifying a term in the dictionary search method:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we search the whole set of variables\n",
    "plain_variablesDict <- hpds::find.in.dictionary(resource, \"\") %>% \n",
    "hpds::extract.entries()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 9248    9\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>patientCount</th><th scope=col>categorical</th><th scope=col>observationCount</th><th scope=col>name</th><th scope=col>min</th><th scope=col>max</th><th scope=col>HpdsDataType</th><th scope=col>categoryValues</th><th scope=col>description</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;list&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>1</td><td>TRUE</td><td>1</td><td>\\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\Phenotypic abnormality\\Abnormality of the nervous system\\Abnormality of nervous system physiology\\Seizures\\Generalized seizures\\Absence seizures\\Typical absence seizures\\                                                                                                                                                                                                                                                                                                                                      </td><td>NA</td><td>NA</td><td>phenotypes</td><td>Positive</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>1</td><td>TRUE</td><td>1</td><td>\\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\Phenotypic abnormality\\Abnormality of the skeletal system\\Abnormality of skeletal morphology\\Abnormal appendicular skeleton morphology\\Abnormality of limb bone morphology\\Abnormality of limb epiphysis morphology\\Abnormality of upper limb epiphysis morphology\\Abnormality of the epiphyses of the hand\\Abnormality of the epiphyses of the phalanges of the hand\\Abnormality of the epiphyses of the middle phalanges of the hand\\Abnormality of the epiphysis of the middle phalanx of the 2nd finger\\</td><td>NA</td><td>NA</td><td>phenotypes</td><td>Positive</td><td>NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 9\n",
       "\\begin{tabular}{r|lllllllll}\n",
       "  & patientCount & categorical & observationCount & name & min & max & HpdsDataType & categoryValues & description\\\\\n",
       "  & <int> & <lgl> & <int> & <chr> & <dbl> & <dbl> & <chr> & <list> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & 1 & TRUE & 1 & \\textbackslash{}04\\_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\textbackslash{}Phenotypic abnormality\\textbackslash{}Abnormality of the nervous system\\textbackslash{}Abnormality of nervous system physiology\\textbackslash{}Seizures\\textbackslash{}Generalized seizures\\textbackslash{}Absence seizures\\textbackslash{}Typical absence seizures\\textbackslash{}                                                                                                                                                                                                                                                                                                                                       & NA & NA & phenotypes & Positive & NA\\\\\n",
       "\t2 & 1 & TRUE & 1 & \\textbackslash{}04\\_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\textbackslash{}Phenotypic abnormality\\textbackslash{}Abnormality of the skeletal system\\textbackslash{}Abnormality of skeletal morphology\\textbackslash{}Abnormal appendicular skeleton morphology\\textbackslash{}Abnormality of limb bone morphology\\textbackslash{}Abnormality of limb epiphysis morphology\\textbackslash{}Abnormality of upper limb epiphysis morphology\\textbackslash{}Abnormality of the epiphyses of the hand\\textbackslash{}Abnormality of the epiphyses of the phalanges of the hand\\textbackslash{}Abnormality of the epiphyses of the middle phalanges of the hand\\textbackslash{}Abnormality of the epiphysis of the middle phalanx of the 2nd finger\\textbackslash{} & NA & NA & phenotypes & Positive & NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 9\n",
       "\n",
       "| <!--/--> | patientCount &lt;int&gt; | categorical &lt;lgl&gt; | observationCount &lt;int&gt; | name &lt;chr&gt; | min &lt;dbl&gt; | max &lt;dbl&gt; | HpdsDataType &lt;chr&gt; | categoryValues &lt;list&gt; | description &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 1 | TRUE | 1 | \\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\Phenotypic abnormality\\Abnormality of the nervous system\\Abnormality of nervous system physiology\\Seizures\\Generalized seizures\\Absence seizures\\Typical absence seizures\\                                                                                                                                                                                                                                                                                                                                       | NA | NA | phenotypes | Positive | NA |\n",
       "| 2 | 1 | TRUE | 1 | \\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\Phenotypic abnormality\\Abnormality of the skeletal system\\Abnormality of skeletal morphology\\Abnormal appendicular skeleton morphology\\Abnormality of limb bone morphology\\Abnormality of limb epiphysis morphology\\Abnormality of upper limb epiphysis morphology\\Abnormality of the epiphyses of the hand\\Abnormality of the epiphyses of the phalanges of the hand\\Abnormality of the epiphyses of the middle phalanges of the hand\\Abnormality of the epiphysis of the middle phalanx of the 2nd finger\\ | NA | NA | phenotypes | Positive | NA |\n",
       "\n"
      ],
      "text/plain": [
       "  patientCount categorical observationCount\n",
       "1 1            TRUE        1               \n",
       "2 1            TRUE        1               \n",
       "  name                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \n",
       "1 \\\\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\\\Phenotypic abnormality\\\\Abnormality of the nervous system\\\\Abnormality of nervous system physiology\\\\Seizures\\\\Generalized seizures\\\\Absence seizures\\\\Typical absence seizures\\\\                                                                                                                                                                                                                                                                                                                                      \n",
       "2 \\\\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\\\Phenotypic abnormality\\\\Abnormality of the skeletal system\\\\Abnormality of skeletal morphology\\\\Abnormal appendicular skeleton morphology\\\\Abnormality of limb bone morphology\\\\Abnormality of limb epiphysis morphology\\\\Abnormality of upper limb epiphysis morphology\\\\Abnormality of the epiphyses of the hand\\\\Abnormality of the epiphyses of the phalanges of the hand\\\\Abnormality of the epiphyses of the middle phalanges of the hand\\\\Abnormality of the epiphysis of the middle phalanx of the 2nd finger\\\\\n",
       "  min max HpdsDataType categoryValues description\n",
       "1 NA  NA  phenotypes   Positive       NA         \n",
       "2 NA  NA  phenotypes   Positive       NA         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# description of the whole dictionary of variables\n",
    "print(dim(plain_variablesDict))\n",
    "head(plain_variablesDict, n = 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The UDN network resource contains 13414 variables described by 11 data fields:\n",
    "* name\n",
    "* HpdsDataType\n",
    "* description\n",
    "* categorical\n",
    "* categoryValues\n",
    "* values\n",
    "* continuous\n",
    "* min\n",
    "* max\n",
    "* observationCount\n",
    "* patientCount\n",
    "\n",
    "The dictionary provides various information about the variables, such as:\n",
    "\n",
    "* observationCount: number of entries with non-null value\n",
    "* categorical: type of the variables, True if categorical, False if continuous/numerical\n",
    "* min/max: only provided for non-categorical variables\n",
    "* HpdsDataType: 'phenotypes' or 'genotypes'. Currently, the API only expsoses'phenotypes' variables\n",
    "\n",
    "Hence, it enables us to:\n",
    "\n",
    "* Use the various variables information as criteria for variable selection.\n",
    "* Use the row names of the DataFrame to get the actual variables names, to be used in the query, as shown below.\n",
    " \n",
    "Variable names (`name` **column** in the dataframe), as currently implemented in the API, aren't straightforward to use because:\n",
    "\n",
    "1. Very long\n",
    "2. Presence of backslashes that requires modification right after copy-pasting.\n",
    "\n",
    "However, using the dictionary to select variables can help to deal with this. \n",
    "\n",
    "##### Parsing variable names\n",
    "We can use an utils function, `get_multiIndex()`, defined in R_lib/utils.R, to add a little more information and ease working with variables names.\n",
    "\n",
    "Although not an official feature of the API, such functionality illustrates how to quickly scan and select groups of related variables.\n",
    "\n",
    "Printing part of the \"parsed names\" Dictionary allows to quickly see the tree-like organisation of the variables. Moreover, original and simplified variable names are now stored respectively in the \"name\" and \"simplified_name\" columns (simplified variable names is simply the last component of the variable name, which usually makes the most sense to know what each variable is about)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 2 × 23</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>level_0</th><th scope=col>level_1</th><th scope=col>level_2</th><th scope=col>level_3</th><th scope=col>level_4</th><th scope=col>level_5</th><th scope=col>level_6</th><th scope=col>level_7</th><th scope=col>level_8</th><th scope=col>level_9</th><th scope=col>⋯</th><th scope=col>level_13</th><th scope=col>level_14</th><th scope=col>simplified_name</th><th scope=col>name</th><th scope=col>observationCount</th><th scope=col>categorical</th><th scope=col>categoryValues</th><th scope=col>min</th><th scope=col>max</th><th scope=col>HpdsDataType</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;list&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>04_Clinical symptoms and physical findings (in HPO, from PhenoTips)</td><td>Phenotypic abnormality</td><td>Abnormality of the nervous system </td><td>Abnormality of nervous system physiology</td><td>Seizures                                 </td><td>Generalized seizures               </td><td>Absence seizures                        </td><td>Typical absence seizures                      </td><td>NA                                      </td><td>NA                                                       </td><td>⋯</td><td>NA</td><td>NA</td><td>Typical absence seizures                                            </td><td>\\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\Phenotypic abnormality\\Abnormality of the nervous system\\Abnormality of nervous system physiology\\Seizures\\Generalized seizures\\Absence seizures\\Typical absence seizures\\                                                                                                                                                                                                                                                                                                                                      </td><td>1</td><td>TRUE</td><td>Positive</td><td>NA</td><td>NA</td><td>phenotypes</td></tr>\n",
       "\t<tr><td>04_Clinical symptoms and physical findings (in HPO, from PhenoTips)</td><td>Phenotypic abnormality</td><td>Abnormality of the skeletal system</td><td>Abnormality of skeletal morphology      </td><td>Abnormal appendicular skeleton morphology</td><td>Abnormality of limb bone morphology</td><td>Abnormality of limb epiphysis morphology</td><td>Abnormality of upper limb epiphysis morphology</td><td>Abnormality of the epiphyses of the hand</td><td>Abnormality of the epiphyses of the phalanges of the hand</td><td>⋯</td><td>NA</td><td>NA</td><td>Abnormality of the epiphysis of the middle phalanx of the 2nd finger</td><td>\\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\Phenotypic abnormality\\Abnormality of the skeletal system\\Abnormality of skeletal morphology\\Abnormal appendicular skeleton morphology\\Abnormality of limb bone morphology\\Abnormality of limb epiphysis morphology\\Abnormality of upper limb epiphysis morphology\\Abnormality of the epiphyses of the hand\\Abnormality of the epiphyses of the phalanges of the hand\\Abnormality of the epiphyses of the middle phalanges of the hand\\Abnormality of the epiphysis of the middle phalanx of the 2nd finger\\</td><td>1</td><td>TRUE</td><td>Positive</td><td>NA</td><td>NA</td><td>phenotypes</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 2 × 23\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " level\\_0 & level\\_1 & level\\_2 & level\\_3 & level\\_4 & level\\_5 & level\\_6 & level\\_7 & level\\_8 & level\\_9 & ⋯ & level\\_13 & level\\_14 & simplified\\_name & name & observationCount & categorical & categoryValues & min & max & HpdsDataType\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & ⋯ & <chr> & <chr> & <chr> & <chr> & <int> & <lgl> & <list> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t 04\\_Clinical symptoms and physical findings (in HPO, from PhenoTips) & Phenotypic abnormality & Abnormality of the nervous system  & Abnormality of nervous system physiology & Seizures                                  & Generalized seizures                & Absence seizures                         & Typical absence seizures                       & NA                                       & NA                                                        & ⋯ & NA & NA & Typical absence seizures                                             & \\textbackslash{}04\\_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\textbackslash{}Phenotypic abnormality\\textbackslash{}Abnormality of the nervous system\\textbackslash{}Abnormality of nervous system physiology\\textbackslash{}Seizures\\textbackslash{}Generalized seizures\\textbackslash{}Absence seizures\\textbackslash{}Typical absence seizures\\textbackslash{}                                                                                                                                                                                                                                                                                                                                       & 1 & TRUE & Positive & NA & NA & phenotypes\\\\\n",
       "\t 04\\_Clinical symptoms and physical findings (in HPO, from PhenoTips) & Phenotypic abnormality & Abnormality of the skeletal system & Abnormality of skeletal morphology       & Abnormal appendicular skeleton morphology & Abnormality of limb bone morphology & Abnormality of limb epiphysis morphology & Abnormality of upper limb epiphysis morphology & Abnormality of the epiphyses of the hand & Abnormality of the epiphyses of the phalanges of the hand & ⋯ & NA & NA & Abnormality of the epiphysis of the middle phalanx of the 2nd finger & \\textbackslash{}04\\_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\textbackslash{}Phenotypic abnormality\\textbackslash{}Abnormality of the skeletal system\\textbackslash{}Abnormality of skeletal morphology\\textbackslash{}Abnormal appendicular skeleton morphology\\textbackslash{}Abnormality of limb bone morphology\\textbackslash{}Abnormality of limb epiphysis morphology\\textbackslash{}Abnormality of upper limb epiphysis morphology\\textbackslash{}Abnormality of the epiphyses of the hand\\textbackslash{}Abnormality of the epiphyses of the phalanges of the hand\\textbackslash{}Abnormality of the epiphyses of the middle phalanges of the hand\\textbackslash{}Abnormality of the epiphysis of the middle phalanx of the 2nd finger\\textbackslash{} & 1 & TRUE & Positive & NA & NA & phenotypes\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 2 × 23\n",
       "\n",
       "| level_0 &lt;chr&gt; | level_1 &lt;chr&gt; | level_2 &lt;chr&gt; | level_3 &lt;chr&gt; | level_4 &lt;chr&gt; | level_5 &lt;chr&gt; | level_6 &lt;chr&gt; | level_7 &lt;chr&gt; | level_8 &lt;chr&gt; | level_9 &lt;chr&gt; | ⋯ ⋯ | level_13 &lt;chr&gt; | level_14 &lt;chr&gt; | simplified_name &lt;chr&gt; | name &lt;chr&gt; | observationCount &lt;int&gt; | categorical &lt;lgl&gt; | categoryValues &lt;list&gt; | min &lt;dbl&gt; | max &lt;dbl&gt; | HpdsDataType &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 04_Clinical symptoms and physical findings (in HPO, from PhenoTips) | Phenotypic abnormality | Abnormality of the nervous system  | Abnormality of nervous system physiology | Seizures                                  | Generalized seizures                | Absence seizures                         | Typical absence seizures                       | NA                                       | NA                                                        | ⋯ | NA | NA | Typical absence seizures                                             | \\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\Phenotypic abnormality\\Abnormality of the nervous system\\Abnormality of nervous system physiology\\Seizures\\Generalized seizures\\Absence seizures\\Typical absence seizures\\                                                                                                                                                                                                                                                                                                                                       | 1 | TRUE | Positive | NA | NA | phenotypes |\n",
       "| 04_Clinical symptoms and physical findings (in HPO, from PhenoTips) | Phenotypic abnormality | Abnormality of the skeletal system | Abnormality of skeletal morphology       | Abnormal appendicular skeleton morphology | Abnormality of limb bone morphology | Abnormality of limb epiphysis morphology | Abnormality of upper limb epiphysis morphology | Abnormality of the epiphyses of the hand | Abnormality of the epiphyses of the phalanges of the hand | ⋯ | NA | NA | Abnormality of the epiphysis of the middle phalanx of the 2nd finger | \\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\Phenotypic abnormality\\Abnormality of the skeletal system\\Abnormality of skeletal morphology\\Abnormal appendicular skeleton morphology\\Abnormality of limb bone morphology\\Abnormality of limb epiphysis morphology\\Abnormality of upper limb epiphysis morphology\\Abnormality of the epiphyses of the hand\\Abnormality of the epiphyses of the phalanges of the hand\\Abnormality of the epiphyses of the middle phalanges of the hand\\Abnormality of the epiphysis of the middle phalanx of the 2nd finger\\ | 1 | TRUE | Positive | NA | NA | phenotypes |\n",
       "\n"
      ],
      "text/plain": [
       "  level_0                                                            \n",
       "1 04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\n",
       "2 04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\n",
       "  level_1                level_2                           \n",
       "1 Phenotypic abnormality Abnormality of the nervous system \n",
       "2 Phenotypic abnormality Abnormality of the skeletal system\n",
       "  level_3                                 \n",
       "1 Abnormality of nervous system physiology\n",
       "2 Abnormality of skeletal morphology      \n",
       "  level_4                                   level_5                            \n",
       "1 Seizures                                  Generalized seizures               \n",
       "2 Abnormal appendicular skeleton morphology Abnormality of limb bone morphology\n",
       "  level_6                                 \n",
       "1 Absence seizures                        \n",
       "2 Abnormality of limb epiphysis morphology\n",
       "  level_7                                       \n",
       "1 Typical absence seizures                      \n",
       "2 Abnormality of upper limb epiphysis morphology\n",
       "  level_8                                 \n",
       "1 NA                                      \n",
       "2 Abnormality of the epiphyses of the hand\n",
       "  level_9                                                   ⋯ level_13 level_14\n",
       "1 NA                                                        ⋯ NA       NA      \n",
       "2 Abnormality of the epiphyses of the phalanges of the hand ⋯ NA       NA      \n",
       "  simplified_name                                                     \n",
       "1 Typical absence seizures                                            \n",
       "2 Abnormality of the epiphysis of the middle phalanx of the 2nd finger\n",
       "  name                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \n",
       "1 \\\\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\\\Phenotypic abnormality\\\\Abnormality of the nervous system\\\\Abnormality of nervous system physiology\\\\Seizures\\\\Generalized seizures\\\\Absence seizures\\\\Typical absence seizures\\\\                                                                                                                                                                                                                                                                                                                                      \n",
       "2 \\\\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\\\Phenotypic abnormality\\\\Abnormality of the skeletal system\\\\Abnormality of skeletal morphology\\\\Abnormal appendicular skeleton morphology\\\\Abnormality of limb bone morphology\\\\Abnormality of limb epiphysis morphology\\\\Abnormality of upper limb epiphysis morphology\\\\Abnormality of the epiphyses of the hand\\\\Abnormality of the epiphyses of the phalanges of the hand\\\\Abnormality of the epiphyses of the middle phalanges of the hand\\\\Abnormality of the epiphysis of the middle phalanx of the 2nd finger\\\\\n",
       "  observationCount categorical categoryValues min max HpdsDataType\n",
       "1 1                TRUE        Positive       NA  NA  phenotypes  \n",
       "2 1                TRUE        Positive       NA  NA  phenotypes  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Display the variables tree hierarchy from the variables name\n",
    "variablesDict <- get_multiIndex_variablesDict(plain_variablesDict)\n",
    "head(variablesDict, n = 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below is a simple example to illustrate the ease of use of a multiIndex dictionary. Let's say we are interested in filtering variables related to \"aplasias\" in the \"nervous system\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 35 23\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 2 × 23</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>level_0</th><th scope=col>level_1</th><th scope=col>level_2</th><th scope=col>level_3</th><th scope=col>level_4</th><th scope=col>level_5</th><th scope=col>level_6</th><th scope=col>level_7</th><th scope=col>level_8</th><th scope=col>level_9</th><th scope=col>⋯</th><th scope=col>level_13</th><th scope=col>level_14</th><th scope=col>simplified_name</th><th scope=col>name</th><th scope=col>observationCount</th><th scope=col>categorical</th><th scope=col>categoryValues</th><th scope=col>min</th><th scope=col>max</th><th scope=col>HpdsDataType</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;list&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>04_Clinical symptoms and physical findings (in HPO, from PhenoTips)</td><td>Phenotypic abnormality</td><td>Abnormality of the nervous system</td><td>Abnormality of nervous system morphology</td><td>Morphological abnormality of the central nervous system</td><td>Abnormality of brain morphology</td><td>Abnormality of forebrain morphology</td><td>Abnormality of the cerebrum     </td><td>Aplasia/Hypoplasia of the cerebrum</td><td>Anencephaly            </td><td>⋯</td><td>NA                       </td><td>NA</td><td>Arrhinencephaly          </td><td>\\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\Phenotypic abnormality\\Abnormality of the nervous system\\Abnormality of nervous system morphology\\Morphological abnormality of the central nervous system\\Abnormality of brain morphology\\Abnormality of forebrain morphology\\Abnormality of the cerebrum\\Aplasia/Hypoplasia of the cerebrum\\Anencephaly\\Arrhinencephaly\\                                                                                                                                       </td><td>1</td><td>TRUE</td><td>Positive</td><td>NA</td><td>NA</td><td>phenotypes</td></tr>\n",
       "\t<tr><td>04_Clinical symptoms and physical findings (in HPO, from PhenoTips)</td><td>Phenotypic abnormality</td><td>Abnormality of the nervous system</td><td>Abnormality of nervous system morphology</td><td>Morphological abnormality of the central nervous system</td><td>Abnormality of brain morphology</td><td>Abnormality of hindbrain morphology</td><td>Abnormality of the metencephalon</td><td>Abnormality of the cerebellum     </td><td>Cerebellar malformation</td><td>⋯</td><td>Dandy-Walker malformation</td><td>NA</td><td>Dandy-Walker malformation</td><td>\\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\Phenotypic abnormality\\Abnormality of the nervous system\\Abnormality of nervous system morphology\\Morphological abnormality of the central nervous system\\Abnormality of brain morphology\\Abnormality of hindbrain morphology\\Abnormality of the metencephalon\\Abnormality of the cerebellum\\Cerebellar malformation\\Abnormality of the cerebellar vermis\\Aplasia/Hypoplasia of the cerebellar vermis\\Cerebellar vermis hypoplasia\\Dandy-Walker malformation\\</td><td>8</td><td>TRUE</td><td>Positive</td><td>NA</td><td>NA</td><td>phenotypes</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 2 × 23\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " level\\_0 & level\\_1 & level\\_2 & level\\_3 & level\\_4 & level\\_5 & level\\_6 & level\\_7 & level\\_8 & level\\_9 & ⋯ & level\\_13 & level\\_14 & simplified\\_name & name & observationCount & categorical & categoryValues & min & max & HpdsDataType\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & ⋯ & <chr> & <chr> & <chr> & <chr> & <int> & <lgl> & <list> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t 04\\_Clinical symptoms and physical findings (in HPO, from PhenoTips) & Phenotypic abnormality & Abnormality of the nervous system & Abnormality of nervous system morphology & Morphological abnormality of the central nervous system & Abnormality of brain morphology & Abnormality of forebrain morphology & Abnormality of the cerebrum      & Aplasia/Hypoplasia of the cerebrum & Anencephaly             & ⋯ & NA                        & NA & Arrhinencephaly           & \\textbackslash{}04\\_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\textbackslash{}Phenotypic abnormality\\textbackslash{}Abnormality of the nervous system\\textbackslash{}Abnormality of nervous system morphology\\textbackslash{}Morphological abnormality of the central nervous system\\textbackslash{}Abnormality of brain morphology\\textbackslash{}Abnormality of forebrain morphology\\textbackslash{}Abnormality of the cerebrum\\textbackslash{}Aplasia/Hypoplasia of the cerebrum\\textbackslash{}Anencephaly\\textbackslash{}Arrhinencephaly\\textbackslash{}                                                                                                                                        & 1 & TRUE & Positive & NA & NA & phenotypes\\\\\n",
       "\t 04\\_Clinical symptoms and physical findings (in HPO, from PhenoTips) & Phenotypic abnormality & Abnormality of the nervous system & Abnormality of nervous system morphology & Morphological abnormality of the central nervous system & Abnormality of brain morphology & Abnormality of hindbrain morphology & Abnormality of the metencephalon & Abnormality of the cerebellum      & Cerebellar malformation & ⋯ & Dandy-Walker malformation & NA & Dandy-Walker malformation & \\textbackslash{}04\\_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\textbackslash{}Phenotypic abnormality\\textbackslash{}Abnormality of the nervous system\\textbackslash{}Abnormality of nervous system morphology\\textbackslash{}Morphological abnormality of the central nervous system\\textbackslash{}Abnormality of brain morphology\\textbackslash{}Abnormality of hindbrain morphology\\textbackslash{}Abnormality of the metencephalon\\textbackslash{}Abnormality of the cerebellum\\textbackslash{}Cerebellar malformation\\textbackslash{}Abnormality of the cerebellar vermis\\textbackslash{}Aplasia/Hypoplasia of the cerebellar vermis\\textbackslash{}Cerebellar vermis hypoplasia\\textbackslash{}Dandy-Walker malformation\\textbackslash{} & 8 & TRUE & Positive & NA & NA & phenotypes\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 2 × 23\n",
       "\n",
       "| level_0 &lt;chr&gt; | level_1 &lt;chr&gt; | level_2 &lt;chr&gt; | level_3 &lt;chr&gt; | level_4 &lt;chr&gt; | level_5 &lt;chr&gt; | level_6 &lt;chr&gt; | level_7 &lt;chr&gt; | level_8 &lt;chr&gt; | level_9 &lt;chr&gt; | ⋯ ⋯ | level_13 &lt;chr&gt; | level_14 &lt;chr&gt; | simplified_name &lt;chr&gt; | name &lt;chr&gt; | observationCount &lt;int&gt; | categorical &lt;lgl&gt; | categoryValues &lt;list&gt; | min &lt;dbl&gt; | max &lt;dbl&gt; | HpdsDataType &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 04_Clinical symptoms and physical findings (in HPO, from PhenoTips) | Phenotypic abnormality | Abnormality of the nervous system | Abnormality of nervous system morphology | Morphological abnormality of the central nervous system | Abnormality of brain morphology | Abnormality of forebrain morphology | Abnormality of the cerebrum      | Aplasia/Hypoplasia of the cerebrum | Anencephaly             | ⋯ | NA                        | NA | Arrhinencephaly           | \\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\Phenotypic abnormality\\Abnormality of the nervous system\\Abnormality of nervous system morphology\\Morphological abnormality of the central nervous system\\Abnormality of brain morphology\\Abnormality of forebrain morphology\\Abnormality of the cerebrum\\Aplasia/Hypoplasia of the cerebrum\\Anencephaly\\Arrhinencephaly\\                                                                                                                                        | 1 | TRUE | Positive | NA | NA | phenotypes |\n",
       "| 04_Clinical symptoms and physical findings (in HPO, from PhenoTips) | Phenotypic abnormality | Abnormality of the nervous system | Abnormality of nervous system morphology | Morphological abnormality of the central nervous system | Abnormality of brain morphology | Abnormality of hindbrain morphology | Abnormality of the metencephalon | Abnormality of the cerebellum      | Cerebellar malformation | ⋯ | Dandy-Walker malformation | NA | Dandy-Walker malformation | \\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\Phenotypic abnormality\\Abnormality of the nervous system\\Abnormality of nervous system morphology\\Morphological abnormality of the central nervous system\\Abnormality of brain morphology\\Abnormality of hindbrain morphology\\Abnormality of the metencephalon\\Abnormality of the cerebellum\\Cerebellar malformation\\Abnormality of the cerebellar vermis\\Aplasia/Hypoplasia of the cerebellar vermis\\Cerebellar vermis hypoplasia\\Dandy-Walker malformation\\ | 8 | TRUE | Positive | NA | NA | phenotypes |\n",
       "\n"
      ],
      "text/plain": [
       "  level_0                                                            \n",
       "1 04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\n",
       "2 04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\n",
       "  level_1                level_2                          \n",
       "1 Phenotypic abnormality Abnormality of the nervous system\n",
       "2 Phenotypic abnormality Abnormality of the nervous system\n",
       "  level_3                                 \n",
       "1 Abnormality of nervous system morphology\n",
       "2 Abnormality of nervous system morphology\n",
       "  level_4                                                \n",
       "1 Morphological abnormality of the central nervous system\n",
       "2 Morphological abnormality of the central nervous system\n",
       "  level_5                         level_6                            \n",
       "1 Abnormality of brain morphology Abnormality of forebrain morphology\n",
       "2 Abnormality of brain morphology Abnormality of hindbrain morphology\n",
       "  level_7                          level_8                           \n",
       "1 Abnormality of the cerebrum      Aplasia/Hypoplasia of the cerebrum\n",
       "2 Abnormality of the metencephalon Abnormality of the cerebellum     \n",
       "  level_9                 ⋯ level_13                  level_14\n",
       "1 Anencephaly             ⋯ NA                        NA      \n",
       "2 Cerebellar malformation ⋯ Dandy-Walker malformation NA      \n",
       "  simplified_name          \n",
       "1 Arrhinencephaly          \n",
       "2 Dandy-Walker malformation\n",
       "  name                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             \n",
       "1 \\\\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\\\Phenotypic abnormality\\\\Abnormality of the nervous system\\\\Abnormality of nervous system morphology\\\\Morphological abnormality of the central nervous system\\\\Abnormality of brain morphology\\\\Abnormality of forebrain morphology\\\\Abnormality of the cerebrum\\\\Aplasia/Hypoplasia of the cerebrum\\\\Anencephaly\\\\Arrhinencephaly\\\\                                                                                                                                       \n",
       "2 \\\\04_Clinical symptoms and physical findings (in HPO, from PhenoTips)\\\\Phenotypic abnormality\\\\Abnormality of the nervous system\\\\Abnormality of nervous system morphology\\\\Morphological abnormality of the central nervous system\\\\Abnormality of brain morphology\\\\Abnormality of hindbrain morphology\\\\Abnormality of the metencephalon\\\\Abnormality of the cerebellum\\\\Cerebellar malformation\\\\Abnormality of the cerebellar vermis\\\\Aplasia/Hypoplasia of the cerebellar vermis\\\\Cerebellar vermis hypoplasia\\\\Dandy-Walker malformation\\\\\n",
       "  observationCount categorical categoryValues min max HpdsDataType\n",
       "1 1                TRUE        Positive       NA  NA  phenotypes  \n",
       "2 8                TRUE        Positive       NA  NA  phenotypes  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mask_system <- variablesDict[,3] == \"Abnormality of the nervous system\"\n",
    "mask_abnormality <- grepl(\"Aplasia\", variablesDict[[\"name\"]])\n",
    "filtered_variables <- variablesDict[mask_system & mask_abnormality,]\n",
    "print(dim(filtered_variables))\n",
    "head(filtered_variables, n = 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Although pretty simple, it can be easily combined with other filters to quickly select necessary variables.\n",
    "\n",
    "#### Explore patient records using _query_\n",
    "\n",
    "Beside from the dictionary, the second cornerstone of the API are the query methods (`hpds::query.select`, `hpds::query.require`, `hpds::query.anyof`, `hpds::query.filter`). They are the entering point to **query and retrieve data from the resource**.\n",
    "\n",
    "First, we need to create a query object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a query object for the resource\n",
    "my_query <- hpds::new.query(resource = resource)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The query object created will be then passed to the different query methods to build the query:  <font color='orange'>hpds::query.select.add(), hpds::query.require.add(), hpds::query.anyof.add(), and hpds::query.filter.add()</font>. Each of those methods accept a query object, a list of variable names, and eventual additional parameters.\n",
    "\n",
    "| Method | Arguments / Input | Output|\n",
    "|--------|-------------------|-------|\n",
    "| query.select.add() | variable names (string) or list of strings | all variables included in the list (no record subsetting)|\n",
    "| query.require.add() | variable names (string) or list of strings | all variables; only records that do not contain null values for input variables |\n",
    "| query.anyof.add() | variable names (string) or list of strings | all variables; only records that contain at least one non-null value for input variables |\n",
    "| query.filter.add() | variable name and additional filtering values | input variable; only records that match filter criteria |\n",
    "\n",
    "All those 4 methods can be combined when building a query. The record eventually returned by the query have to meet all the different specified filters.\n",
    "\n",
    "##### Building the query\n",
    "Let's say we want to check some demographics about the data in UDN. We will filter to variables that have observation counts > 50% patient counts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"We have found 6 demographics variable(s) which have observation counts > 50% of patient counts (listed below).\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'\\\\00_Demographics\\\\Age at UDN Evaluation (in years)\\\\'</li><li>'\\\\00_Demographics\\\\Age at symptom onset in years\\\\'</li><li>'\\\\00_Demographics\\\\Ethnicity\\\\'</li><li>'\\\\00_Demographics\\\\Gender\\\\'</li><li>'\\\\00_Demographics\\\\Current age in years\\\\Current age in years\\\\'</li><li>'\\\\00_Demographics\\\\Race\\\\'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item '\\textbackslash{}\\textbackslash{}00\\_Demographics\\textbackslash{}\\textbackslash{}Age at UDN Evaluation (in years)\\textbackslash{}\\textbackslash{}'\n",
       "\\item '\\textbackslash{}\\textbackslash{}00\\_Demographics\\textbackslash{}\\textbackslash{}Age at symptom onset in years\\textbackslash{}\\textbackslash{}'\n",
       "\\item '\\textbackslash{}\\textbackslash{}00\\_Demographics\\textbackslash{}\\textbackslash{}Ethnicity\\textbackslash{}\\textbackslash{}'\n",
       "\\item '\\textbackslash{}\\textbackslash{}00\\_Demographics\\textbackslash{}\\textbackslash{}Gender\\textbackslash{}\\textbackslash{}'\n",
       "\\item '\\textbackslash{}\\textbackslash{}00\\_Demographics\\textbackslash{}\\textbackslash{}Current age in years\\textbackslash{}\\textbackslash{}Current age in years\\textbackslash{}\\textbackslash{}'\n",
       "\\item '\\textbackslash{}\\textbackslash{}00\\_Demographics\\textbackslash{}\\textbackslash{}Race\\textbackslash{}\\textbackslash{}'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. '\\\\00_Demographics\\\\Age at UDN Evaluation (in years)\\\\'\n",
       "2. '\\\\00_Demographics\\\\Age at symptom onset in years\\\\'\n",
       "3. '\\\\00_Demographics\\\\Ethnicity\\\\'\n",
       "4. '\\\\00_Demographics\\\\Gender\\\\'\n",
       "5. '\\\\00_Demographics\\\\Current age in years\\\\Current age in years\\\\'\n",
       "6. '\\\\00_Demographics\\\\Race\\\\'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"\\\\00_Demographics\\\\Age at UDN Evaluation (in years)\\\\\"          \n",
       "[2] \"\\\\00_Demographics\\\\Age at symptom onset in years\\\\\"             \n",
       "[3] \"\\\\00_Demographics\\\\Ethnicity\\\\\"                                 \n",
       "[4] \"\\\\00_Demographics\\\\Gender\\\\\"                                    \n",
       "[5] \"\\\\00_Demographics\\\\Current age in years\\\\Current age in years\\\\\"\n",
       "[6] \"\\\\00_Demographics\\\\Race\\\\\"                                      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# select demographic variable names\n",
    "demographicsDict <- hpds::find.in.dictionary(resource, \"demographics\") %>% \n",
    "    hpds::extract.entries()\n",
    "mask_obs <- demographicsDict %>% filter(observationCount > patientCount * 0.50)\n",
    "selected_varnames <- mask_obs %>% pull(name) \n",
    "print(paste0('We have found ', length(selected_varnames), ' demographics variable(s) which have observation counts > 50% of patient counts (listed below).'))\n",
    "selected_varnames"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You may warning messages containing the following text when building your query with multiple variables: \n",
    "“the condition has length > 1 and only the first element will be used” - this can be ignored.\n",
    "\n",
    "To double check that your filter has been applied to your query, you can run ```hpds::query.show(query = my_query)```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in if (has.key(key, self$data) == TRUE) {:\n",
      "“the condition has length > 1 and only the first element will be used”\n",
      "Warning message in if (str_detect(norm_str, \"rs[0-9]+$\")) {:\n",
      "“the condition has length > 1 and only the first element will be used”\n",
      "Warning message in if (str_detect(norm_str, \"[0-9]+,[0-9\\\\.]+,.*\")) {:\n",
      "“the condition has length > 1 and only the first element will be used”\n"
     ]
    }
   ],
   "source": [
    "# build and query for demographics patient data\n",
    "hpds::query.select.add(query=my_query, keys=selected_varnames)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Retrieving the data\n",
    "Once our query object is  built, we use the `query.run()` method to retrieve the data corresponding to our query."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# retrieve the query result as a dataframe\n",
    "demographics_data <- hpds::query.run(my_query, result.type=\"dataframe\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 2048    7\n"
     ]
    }
   ],
   "source": [
    "print(dim(demographics_data))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Patient ID</th><th scope=col>\\00_Demographics\\Age at UDN Evaluation (in years)\\</th><th scope=col>\\00_Demographics\\Age at symptom onset in years\\</th><th scope=col>\\00_Demographics\\Current age in years\\Current age in years\\</th><th scope=col>\\00_Demographics\\Ethnicity\\</th><th scope=col>\\00_Demographics\\Gender\\</th><th scope=col>\\00_Demographics\\Race\\</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td> 6</td><td>25</td><td>11</td><td>30</td><td>Not Hispanic or Latino</td><td>Male  </td><td>Asian                    </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>16</td><td> 9</td><td> 0</td><td>11</td><td>Not Hispanic or Latino</td><td>Male  </td><td>White                    </td></tr>\n",
       "\t<tr><th scope=row>3</th><td>17</td><td>21</td><td> 6</td><td>24</td><td>Hispanic or Latino    </td><td>Female</td><td>White                    </td></tr>\n",
       "\t<tr><th scope=row>4</th><td>20</td><td> 9</td><td> 4</td><td>12</td><td>Not Hispanic or Latino</td><td>Male  </td><td>Black or African American</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>21</td><td> 7</td><td> 0</td><td> 9</td><td>Hispanic or Latino    </td><td>Female</td><td>White                    </td></tr>\n",
       "\t<tr><th scope=row>6</th><td>24</td><td> 5</td><td> 0</td><td>10</td><td>Not Hispanic or Latino</td><td>Male  </td><td>White                    </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & Patient ID & \\textbackslash{}00\\_Demographics\\textbackslash{}Age at UDN Evaluation (in years)\\textbackslash{} & \\textbackslash{}00\\_Demographics\\textbackslash{}Age at symptom onset in years\\textbackslash{} & \\textbackslash{}00\\_Demographics\\textbackslash{}Current age in years\\textbackslash{}Current age in years\\textbackslash{} & \\textbackslash{}00\\_Demographics\\textbackslash{}Ethnicity\\textbackslash{} & \\textbackslash{}00\\_Demographics\\textbackslash{}Gender\\textbackslash{} & \\textbackslash{}00\\_Demographics\\textbackslash{}Race\\textbackslash{}\\\\\n",
       "  & <int> & <dbl> & <dbl> & <dbl> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 &  6 & 25 & 11 & 30 & Not Hispanic or Latino & Male   & Asian                    \\\\\n",
       "\t2 & 16 &  9 &  0 & 11 & Not Hispanic or Latino & Male   & White                    \\\\\n",
       "\t3 & 17 & 21 &  6 & 24 & Hispanic or Latino     & Female & White                    \\\\\n",
       "\t4 & 20 &  9 &  4 & 12 & Not Hispanic or Latino & Male   & Black or African American\\\\\n",
       "\t5 & 21 &  7 &  0 &  9 & Hispanic or Latino     & Female & White                    \\\\\n",
       "\t6 & 24 &  5 &  0 & 10 & Not Hispanic or Latino & Male   & White                    \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 7\n",
       "\n",
       "| <!--/--> | Patient ID &lt;int&gt; | \\00_Demographics\\Age at UDN Evaluation (in years)\\ &lt;dbl&gt; | \\00_Demographics\\Age at symptom onset in years\\ &lt;dbl&gt; | \\00_Demographics\\Current age in years\\Current age in years\\ &lt;dbl&gt; | \\00_Demographics\\Ethnicity\\ &lt;chr&gt; | \\00_Demographics\\Gender\\ &lt;chr&gt; | \\00_Demographics\\Race\\ &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 1 |  6 | 25 | 11 | 30 | Not Hispanic or Latino | Male   | Asian                     |\n",
       "| 2 | 16 |  9 |  0 | 11 | Not Hispanic or Latino | Male   | White                     |\n",
       "| 3 | 17 | 21 |  6 | 24 | Hispanic or Latino     | Female | White                     |\n",
       "| 4 | 20 |  9 |  4 | 12 | Not Hispanic or Latino | Male   | Black or African American |\n",
       "| 5 | 21 |  7 |  0 |  9 | Hispanic or Latino     | Female | White                     |\n",
       "| 6 | 24 |  5 |  0 | 10 | Not Hispanic or Latino | Male   | White                     |\n",
       "\n"
      ],
      "text/plain": [
       "  Patient ID \\\\00_Demographics\\\\Age at UDN Evaluation (in years)\\\\\n",
       "1  6         25                                                   \n",
       "2 16          9                                                   \n",
       "3 17         21                                                   \n",
       "4 20          9                                                   \n",
       "5 21          7                                                   \n",
       "6 24          5                                                   \n",
       "  \\\\00_Demographics\\\\Age at symptom onset in years\\\\\n",
       "1 11                                                \n",
       "2  0                                                \n",
       "3  6                                                \n",
       "4  4                                                \n",
       "5  0                                                \n",
       "6  0                                                \n",
       "  \\\\00_Demographics\\\\Current age in years\\\\Current age in years\\\\\n",
       "1 30                                                             \n",
       "2 11                                                             \n",
       "3 24                                                             \n",
       "4 12                                                             \n",
       "5  9                                                             \n",
       "6 10                                                             \n",
       "  \\\\00_Demographics\\\\Ethnicity\\\\ \\\\00_Demographics\\\\Gender\\\\\n",
       "1 Not Hispanic or Latino         Male                       \n",
       "2 Not Hispanic or Latino         Male                       \n",
       "3 Hispanic or Latino             Female                     \n",
       "4 Not Hispanic or Latino         Male                       \n",
       "5 Hispanic or Latino             Female                     \n",
       "6 Not Hispanic or Latino         Male                       \n",
       "  \\\\00_Demographics\\\\Race\\\\\n",
       "1 Asian                    \n",
       "2 White                    \n",
       "3 White                    \n",
       "4 Black or African American\n",
       "5 White                    \n",
       "6 White                    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(demographics_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Working with Variant Data\n",
    "You can also use the query object to explore variant data. In this example, let's look at variants for the CHD8 gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a new query\n",
    "my_query <- hpds::new.query(resource = resource)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add a filter for a categorical variant: CHD8\n",
    "hpds::query.filter.add(query=my_query, keys=\"Gene_with_variant\", \"CHD8\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Before calling the full data frame of variants, let's ensure that the approximate total count of variants being returned by our query is of a reasonable size. Queries returning more than 100,000 variants could crash your workbook. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in query$getVariantsApproximateCount():\n",
      "“NAs introduced by coercion”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "1805"
      ],
      "text/latex": [
       "1805"
      ],
      "text/markdown": [
       "1805"
      ],
      "text/plain": [
       "[1] 1805"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "variantCount <- hpds::query.run(my_query, result.type=\"variantsApproximateCount\")\n",
    "variantCount\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 912</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>CHROM</th><th scope=col>POSITION</th><th scope=col>REF</th><th scope=col>ALT</th><th scope=col>Variant_consequence_calculated</th><th scope=col>Variant_class</th><th scope=col>Gene_with_variant</th><th scope=col>Variant_severity</th><th scope=col>Variant_frequency_as_text</th><th scope=col>Patients.with.this.variant.in.subset</th><th scope=col>⋯</th><th scope=col>UDN336336</th><th scope=col>UDN906298</th><th scope=col>UDN556501</th><th scope=col>UDN889016</th><th scope=col>UDN679990</th><th scope=col>UDN909615</th><th scope=col>UDN373820</th><th scope=col>UDN343939</th><th scope=col>UDN340901</th><th scope=col>UDN001168</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>14</td><td>21380203</td><td>G  </td><td>A   </td><td>upstream_gene_variant,downstream_gene_variant,intron_variant</td><td>SNV      </td><td>SUPT16H,CHD8,LOC107984643,AL135744.1</td><td>MODIFIER</td><td>Novel</td><td>1/901 </td><td>⋯</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>14</td><td>21380208</td><td>T  </td><td>C   </td><td>upstream_gene_variant,downstream_gene_variant,intron_variant</td><td>SNV      </td><td>SUPT16H,CHD8,LOC107984643,AL135744.1</td><td>MODIFIER</td><td>Novel</td><td>4/901 </td><td>⋯</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>14</td><td>21380286</td><td>G  </td><td>A   </td><td>upstream_gene_variant,downstream_gene_variant,intron_variant</td><td>SNV      </td><td>SUPT16H,CHD8,LOC107984643,AL135744.1</td><td>MODIFIER</td><td>Novel</td><td>1/901 </td><td>⋯</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>14</td><td>21380296</td><td>ATT</td><td>A   </td><td>upstream_gene_variant,downstream_gene_variant,intron_variant</td><td>deletion </td><td>SUPT16H,CHD8,LOC107984643,AL135744.1</td><td>MODIFIER</td><td>Novel</td><td>1/901 </td><td>⋯</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>14</td><td>21380296</td><td>A  </td><td>AT  </td><td>upstream_gene_variant,downstream_gene_variant,intron_variant</td><td>insertion</td><td>SUPT16H,CHD8,LOC107984643,AL135744.1</td><td>MODIFIER</td><td>Novel</td><td>3/901 </td><td>⋯</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>14</td><td>21380296</td><td>A  </td><td>ATTT</td><td>upstream_gene_variant,downstream_gene_variant,intron_variant</td><td>insertion</td><td>SUPT16H,CHD8,LOC107984643,AL135744.1</td><td>MODIFIER</td><td>Novel</td><td>38/901</td><td>⋯</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/0</td><td>0/1</td><td>0/0</td><td>0/0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 912\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & CHROM & POSITION & REF & ALT & Variant\\_consequence\\_calculated & Variant\\_class & Gene\\_with\\_variant & Variant\\_severity & Variant\\_frequency\\_as\\_text & Patients.with.this.variant.in.subset & ⋯ & UDN336336 & UDN906298 & UDN556501 & UDN889016 & UDN679990 & UDN909615 & UDN373820 & UDN343939 & UDN340901 & UDN001168\\\\\n",
       "  & <int> & <int> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & ⋯ & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & 14 & 21380203 & G   & A    & upstream\\_gene\\_variant,downstream\\_gene\\_variant,intron\\_variant & SNV       & SUPT16H,CHD8,LOC107984643,AL135744.1 & MODIFIER & Novel & 1/901  & ⋯ & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0\\\\\n",
       "\t2 & 14 & 21380208 & T   & C    & upstream\\_gene\\_variant,downstream\\_gene\\_variant,intron\\_variant & SNV       & SUPT16H,CHD8,LOC107984643,AL135744.1 & MODIFIER & Novel & 4/901  & ⋯ & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0\\\\\n",
       "\t3 & 14 & 21380286 & G   & A    & upstream\\_gene\\_variant,downstream\\_gene\\_variant,intron\\_variant & SNV       & SUPT16H,CHD8,LOC107984643,AL135744.1 & MODIFIER & Novel & 1/901  & ⋯ & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0\\\\\n",
       "\t4 & 14 & 21380296 & ATT & A    & upstream\\_gene\\_variant,downstream\\_gene\\_variant,intron\\_variant & deletion  & SUPT16H,CHD8,LOC107984643,AL135744.1 & MODIFIER & Novel & 1/901  & ⋯ & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0\\\\\n",
       "\t5 & 14 & 21380296 & A   & AT   & upstream\\_gene\\_variant,downstream\\_gene\\_variant,intron\\_variant & insertion & SUPT16H,CHD8,LOC107984643,AL135744.1 & MODIFIER & Novel & 3/901  & ⋯ & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0\\\\\n",
       "\t6 & 14 & 21380296 & A   & ATTT & upstream\\_gene\\_variant,downstream\\_gene\\_variant,intron\\_variant & insertion & SUPT16H,CHD8,LOC107984643,AL135744.1 & MODIFIER & Novel & 38/901 & ⋯ & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/0 & 0/1 & 0/0 & 0/0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 912\n",
       "\n",
       "| <!--/--> | CHROM &lt;int&gt; | POSITION &lt;int&gt; | REF &lt;chr&gt; | ALT &lt;chr&gt; | Variant_consequence_calculated &lt;chr&gt; | Variant_class &lt;chr&gt; | Gene_with_variant &lt;chr&gt; | Variant_severity &lt;chr&gt; | Variant_frequency_as_text &lt;chr&gt; | Patients.with.this.variant.in.subset &lt;chr&gt; | ⋯ ⋯ | UDN336336 &lt;chr&gt; | UDN906298 &lt;chr&gt; | UDN556501 &lt;chr&gt; | UDN889016 &lt;chr&gt; | UDN679990 &lt;chr&gt; | UDN909615 &lt;chr&gt; | UDN373820 &lt;chr&gt; | UDN343939 &lt;chr&gt; | UDN340901 &lt;chr&gt; | UDN001168 &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 14 | 21380203 | G   | A    | upstream_gene_variant,downstream_gene_variant,intron_variant | SNV       | SUPT16H,CHD8,LOC107984643,AL135744.1 | MODIFIER | Novel | 1/901  | ⋯ | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 |\n",
       "| 2 | 14 | 21380208 | T   | C    | upstream_gene_variant,downstream_gene_variant,intron_variant | SNV       | SUPT16H,CHD8,LOC107984643,AL135744.1 | MODIFIER | Novel | 4/901  | ⋯ | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 |\n",
       "| 3 | 14 | 21380286 | G   | A    | upstream_gene_variant,downstream_gene_variant,intron_variant | SNV       | SUPT16H,CHD8,LOC107984643,AL135744.1 | MODIFIER | Novel | 1/901  | ⋯ | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 |\n",
       "| 4 | 14 | 21380296 | ATT | A    | upstream_gene_variant,downstream_gene_variant,intron_variant | deletion  | SUPT16H,CHD8,LOC107984643,AL135744.1 | MODIFIER | Novel | 1/901  | ⋯ | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 |\n",
       "| 5 | 14 | 21380296 | A   | AT   | upstream_gene_variant,downstream_gene_variant,intron_variant | insertion | SUPT16H,CHD8,LOC107984643,AL135744.1 | MODIFIER | Novel | 3/901  | ⋯ | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 |\n",
       "| 6 | 14 | 21380296 | A   | ATTT | upstream_gene_variant,downstream_gene_variant,intron_variant | insertion | SUPT16H,CHD8,LOC107984643,AL135744.1 | MODIFIER | Novel | 38/901 | ⋯ | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/0 | 0/1 | 0/0 | 0/0 |\n",
       "\n"
      ],
      "text/plain": [
       "  CHROM POSITION REF ALT \n",
       "1 14    21380203 G   A   \n",
       "2 14    21380208 T   C   \n",
       "3 14    21380286 G   A   \n",
       "4 14    21380296 ATT A   \n",
       "5 14    21380296 A   AT  \n",
       "6 14    21380296 A   ATTT\n",
       "  Variant_consequence_calculated                               Variant_class\n",
       "1 upstream_gene_variant,downstream_gene_variant,intron_variant SNV          \n",
       "2 upstream_gene_variant,downstream_gene_variant,intron_variant SNV          \n",
       "3 upstream_gene_variant,downstream_gene_variant,intron_variant SNV          \n",
       "4 upstream_gene_variant,downstream_gene_variant,intron_variant deletion     \n",
       "5 upstream_gene_variant,downstream_gene_variant,intron_variant insertion    \n",
       "6 upstream_gene_variant,downstream_gene_variant,intron_variant insertion    \n",
       "  Gene_with_variant                    Variant_severity\n",
       "1 SUPT16H,CHD8,LOC107984643,AL135744.1 MODIFIER        \n",
       "2 SUPT16H,CHD8,LOC107984643,AL135744.1 MODIFIER        \n",
       "3 SUPT16H,CHD8,LOC107984643,AL135744.1 MODIFIER        \n",
       "4 SUPT16H,CHD8,LOC107984643,AL135744.1 MODIFIER        \n",
       "5 SUPT16H,CHD8,LOC107984643,AL135744.1 MODIFIER        \n",
       "6 SUPT16H,CHD8,LOC107984643,AL135744.1 MODIFIER        \n",
       "  Variant_frequency_as_text Patients.with.this.variant.in.subset ⋯ UDN336336\n",
       "1 Novel                     1/901                                ⋯ 0/0      \n",
       "2 Novel                     4/901                                ⋯ 0/0      \n",
       "3 Novel                     1/901                                ⋯ 0/0      \n",
       "4 Novel                     1/901                                ⋯ 0/0      \n",
       "5 Novel                     3/901                                ⋯ 0/0      \n",
       "6 Novel                     38/901                               ⋯ 0/0      \n",
       "  UDN906298 UDN556501 UDN889016 UDN679990 UDN909615 UDN373820 UDN343939\n",
       "1 0/0       0/0       0/0       0/0       0/0       0/0       0/0      \n",
       "2 0/0       0/0       0/0       0/0       0/0       0/0       0/0      \n",
       "3 0/0       0/0       0/0       0/0       0/0       0/0       0/0      \n",
       "4 0/0       0/0       0/0       0/0       0/0       0/0       0/0      \n",
       "5 0/0       0/0       0/0       0/0       0/0       0/0       0/0      \n",
       "6 0/0       0/0       0/0       0/0       0/0       0/0       0/1      \n",
       "  UDN340901 UDN001168\n",
       "1 0/0       0/0      \n",
       "2 0/0       0/0      \n",
       "3 0/0       0/0      \n",
       "4 0/0       0/0      \n",
       "5 0/0       0/0      \n",
       "6 0/0       0/0      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "variant_data <- hpds::query.run(my_query, result.type=\"variantsDataFrame\")\n",
    "head(variant_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Another example of a genomic filter is looking at the variant frequency.\n",
    "- Novel variants are not found in the rest of the population\n",
    "- Rare variants are found in <1% of the population\n",
    "- Common variants are found in >= 1% of the population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 1 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>description</th><th scope=col>name</th><th scope=col>min</th><th scope=col>categorical</th><th scope=col>patientCount</th><th scope=col>observationCount</th><th scope=col>max</th><th scope=col>HpdsDataType</th><th scope=col>categoryValues</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;list&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Description=\"The variant allele frequency in gnomAD combined population as discrete text categories. Possible values: Novel, Rare (variant frequency less than 1%), Common (variant frequency greater than or equal to 1%).\"</td><td>Variant_frequency_as_text</td><td>NA</td><td>TRUE</td><td>NA</td><td>NA</td><td>NA</td><td>info</td><td>Novel , Rare  , Common</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 1 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " description & name & min & categorical & patientCount & observationCount & max & HpdsDataType & categoryValues\\\\\n",
       " <chr> & <chr> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <chr> & <list>\\\\\n",
       "\\hline\n",
       "\t Description=\"The variant allele frequency in gnomAD combined population as discrete text categories. Possible values: Novel, Rare (variant frequency less than 1\\%), Common (variant frequency greater than or equal to 1\\%).\" & Variant\\_frequency\\_as\\_text & NA & TRUE & NA & NA & NA & info & Novel , Rare  , Common\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 1 × 9\n",
       "\n",
       "| description &lt;chr&gt; | name &lt;chr&gt; | min &lt;lgl&gt; | categorical &lt;lgl&gt; | patientCount &lt;lgl&gt; | observationCount &lt;lgl&gt; | max &lt;lgl&gt; | HpdsDataType &lt;chr&gt; | categoryValues &lt;list&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| Description=\"The variant allele frequency in gnomAD combined population as discrete text categories. Possible values: Novel, Rare (variant frequency less than 1%), Common (variant frequency greater than or equal to 1%).\" | Variant_frequency_as_text | NA | TRUE | NA | NA | NA | info | Novel , Rare  , Common |\n",
       "\n"
      ],
      "text/plain": [
       "  description                                                                                                                                                                                                                 \n",
       "1 Description=\"The variant allele frequency in gnomAD combined population as discrete text categories. Possible values: Novel, Rare (variant frequency less than 1%), Common (variant frequency greater than or equal to 1%).\"\n",
       "  name                      min categorical patientCount observationCount max\n",
       "1 Variant_frequency_as_text NA  TRUE        NA           NA               NA \n",
       "  HpdsDataType categoryValues        \n",
       "1 info         Novel , Rare  , Common"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# what is Variant_frequency_in_ExAC?\n",
    "\n",
    "hpds::find.in.dictionary(resource, \"Variant_frequency\") %>% \n",
    "    hpds::extract.entries()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERROR: HTTP response was bad\n",
      "Response [https://udn.hms.harvard.edu/picsure/query/sync/]\n",
      "  Date: 2021-07-29 21:04\n",
      "  Status: 502\n",
      "  Content-Type: text/html; charset=iso-8859-1\n",
      "  Size: 341 B\n",
      "<!DOCTYPE HTML PUBLIC \"-//IETF//DTD HTML 2.0//EN\">\n",
      "<html><head>\n",
      "<title>502 Proxy Error</title>\n",
      "</head><body>\n",
      "<h1>Proxy Error</h1>\n",
      "<p>The proxy server received an invalid\r\n",
      "response from an upstream server.<br />\r\n",
      "The proxy server could not handle the request<p>Reason: <strong>Error reading...\n",
      "</body></html>\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 0 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>{results:{}</th><th scope=col>error:True}</th></tr>\n",
       "\t<tr><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 0 × 2\n",
       "\\begin{tabular}{ll}\n",
       " \\{results:\\{\\} & error:True\\}\\\\\n",
       " <lgl> & <lgl>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 0 × 2\n",
       "\n",
       "| {results:{} &lt;lgl&gt; | error:True} &lt;lgl&gt; |\n",
       "|---|---|\n",
       "\n"
      ],
      "text/plain": [
       "     {results:{} error:True}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Example querying for novel variants\n",
    "my_query <- hpds::new.query(resource = resource)\n",
    "hpds::query.filter.add(query=my_query, keys=\"Variant_frequency_as_text\")\n",
    "#variant_data <- hpds::query.run(my_query, result.type=\"variantsDataFrame\")\n",
    "variant_data <- hpds::query.run(my_query)\n",
    "\n",
    "head(variant_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we can combined genomic and phenotypic filters into a single query:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ERROR: cannot add, key does not exist in resource: Variant_frequency_in_ExAC\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 0 × 1</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>No.Variants.Found</th></tr>\n",
       "\t<tr><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 0 × 1\n",
       "\\begin{tabular}{l}\n",
       " No.Variants.Found\\\\\n",
       " <lgl>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 0 × 1\n",
       "\n",
       "| No.Variants.Found &lt;lgl&gt; |\n",
       "|---|\n",
       "\n"
      ],
      "text/plain": [
       "     No.Variants.Found"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Example combining variant and phenotype queries\n",
    "my_query <- hpds::new.query(resource = resource)\n",
    "hpds::query.filter.add(query=my_query, keys=\"Variant_frequency_in_text\", 0.4, 0.5)\n",
    "hpds::query.filter.add(query=my_query, keys=\"\\\\00_Demographics\\\\Gender\\\\\", \"Female\")\n",
    "variant_data <- hpds::query.run(my_query, result.type=\"variantsDataFrame\")\n",
    "head(variant_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [],
   "source": [
    "?query.filter.add"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
