{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "statistical-extra",
   "metadata": {},
   "source": [
    "# Identifying and Extracting Longitudinal Variables using python PIC-SURE API"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "useful-spectacular",
   "metadata": {},
   "source": [
    "This tutorial notebook will demonstrate how to idetify and extract longitudinal variables using the R PIC-SURE API. Longitudinal variables are defined as containing multiple 'Exam' or 'Visit' descriptions within their concept path.\n",
    "\n",
    "\n",
    "In this example, we will find the patient level data for a lipid-related longitudinal variable within the Framingham Heart study. We will:\n",
    "1. Identify what longitudinal variables are associated with the keywords of interest (lipid, triglyceride), and how many exams / visits are associated with each one\n",
    "2. Select a longitudinal variable of interest from a specific study (Framingham heart study)\n",
    "3. Extract patient level data into a dataframe where each rows represent patients and columns represent visits"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "studied-native",
   "metadata": {},
   "source": [
    "For a more basic introduction to the python PIC-SURE API, see the `1_PICSURE_API_101.ipynb` notebook.\n",
    " \n",
    "**Before running this notebook, please be sure to get a user-specific security token. For more information about how to proceed, see the \\\"Get your security token\\\" instructions in the [README.md](https://github.com/hms-dbmi/Access-to-Data-using-PIC-SURE-API/tree/harmonized_lipid_measurements_example/NHLBI_BioData_Catalyst#get-your-security-token).**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "maritime-stopping",
   "metadata": {},
   "source": [
    "## Environment Set-Up"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "organic-casino",
   "metadata": {},
   "source": [
    "### System Requirements\n",
    "- python 3.6 or later\n",
    "- pip python package manager, already available in most systems with a python interpreter installed ([pip installation instructions](https://pip.pypa.io/en/stable/installing/))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "retained-abraham",
   "metadata": {},
   "source": [
    "### Install Packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "annoying-sixth",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: numpy>=1.16.4 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from -r requirements.txt (line 1)) (1.19.5)\n",
      "Requirement already satisfied: matplotlib>=3.1.1 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from -r requirements.txt (line 2)) (3.3.4)\n",
      "Requirement already satisfied: pandas>=0.25.3 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from -r requirements.txt (line 3)) (1.1.5)\n",
      "Requirement already satisfied: scipy>=1.3.1 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from -r requirements.txt (line 4)) (1.5.3)\n",
      "Collecting tqdm>=4.38.0\n",
      "  Downloading tqdm-4.60.0-py2.py3-none-any.whl (75 kB)\n",
      "\u001b[K     |████████████████████████████████| 75 kB 1.1 MB/s  eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied: statsmodels>=0.10.2 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from -r requirements.txt (line 6)) (0.12.2)\n",
      "Requirement already satisfied: python-dateutil>=2.1 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from matplotlib>=3.1.1->-r requirements.txt (line 2)) (2.8.1)\n",
      "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.3 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from matplotlib>=3.1.1->-r requirements.txt (line 2)) (2.4.7)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from matplotlib>=3.1.1->-r requirements.txt (line 2)) (1.3.1)\n",
      "Requirement already satisfied: cycler>=0.10 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from matplotlib>=3.1.1->-r requirements.txt (line 2)) (0.10.0)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from matplotlib>=3.1.1->-r requirements.txt (line 2)) (7.2.0)\n",
      "Requirement already satisfied: pytz>=2017.2 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from pandas>=0.25.3->-r requirements.txt (line 3)) (2021.1)\n",
      "Requirement already satisfied: patsy>=0.5 in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from statsmodels>=0.10.2->-r requirements.txt (line 6)) (0.5.1)\n",
      "Requirement already satisfied: six in /home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages (from cycler>=0.10->matplotlib>=3.1.1->-r requirements.txt (line 2)) (1.15.0)\n",
      "Installing collected packages: tqdm\n",
      "Successfully installed tqdm-4.60.0\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "!{sys.executable} -m pip install -r requirements.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "disturbed-scheme",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting git+https://github.com/hms-dbmi/pic-sure-python-client.git\n",
      "  Cloning https://github.com/hms-dbmi/pic-sure-python-client.git to /tmp/pip-req-build-0pc0ivdr\n",
      "  Running command git clone -q https://github.com/hms-dbmi/pic-sure-python-client.git /tmp/pip-req-build-0pc0ivdr\n",
      "Building wheels for collected packages: PicSureClient\n",
      "  Building wheel for PicSureClient (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for PicSureClient: filename=PicSureClient-0.1.0-py2.py3-none-any.whl size=10225 sha256=68757d3d2e0bbd2ed09ca170a59d5d28fc9301106ec808ea3c1ed8f84f568fce\n",
      "  Stored in directory: /tmp/pip-ephem-wheel-cache-37h4ehtv/wheels/31/ef/21/e362bba8de04e0072fafec9f77bd1abdf7e166213d27e98729\n",
      "Successfully built PicSureClient\n",
      "Installing collected packages: PicSureClient\n",
      "Successfully installed PicSureClient-0.1.0\n",
      "Collecting git+https://github.com/hms-dbmi/pic-sure-python-adapter-hpds.git\n",
      "  Cloning https://github.com/hms-dbmi/pic-sure-python-adapter-hpds.git to /tmp/pip-req-build-8sm6jlbh\n",
      "  Running command git clone -q https://github.com/hms-dbmi/pic-sure-python-adapter-hpds.git /tmp/pip-req-build-8sm6jlbh\n",
      "Collecting httplib2\n",
      "  Downloading httplib2-0.19.1-py3-none-any.whl (95 kB)\n",
      "\u001b[K     |████████████████████████████████| 95 kB 5.2 MB/s  eta 0:00:01\n",
      "\u001b[?25hCollecting pyparsing<3,>=2.4.2\n",
      "  Downloading pyparsing-2.4.7-py2.py3-none-any.whl (67 kB)\n",
      "\u001b[K     |████████████████████████████████| 67 kB 4.9 MB/s eta 0:00:011\n",
      "\u001b[?25hBuilding wheels for collected packages: PicSureHpdsLib\n",
      "  Building wheel for PicSureHpdsLib (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for PicSureHpdsLib: filename=PicSureHpdsLib-0.9.0-py2.py3-none-any.whl size=21999 sha256=0bda5c3ed4fe1d4932e050c0395e6ccb2c7d5f02a46603a8c99076b5469c97ee\n",
      "  Stored in directory: /tmp/pip-ephem-wheel-cache-lkh8y0v0/wheels/ae/d9/1a/c8c0ac8151b575c845efddc061fe014d86c51d1fd2c408907c\n",
      "Successfully built PicSureHpdsLib\n",
      "Installing collected packages: pyparsing, httplib2, PicSureHpdsLib\n",
      "  Attempting uninstall: pyparsing\n",
      "    Found existing installation: pyparsing 2.4.7\n",
      "    Uninstalling pyparsing-2.4.7:\n",
      "      Successfully uninstalled pyparsing-2.4.7\n",
      "Successfully installed PicSureHpdsLib-0.9.0 httplib2-0.19.1 pyparsing-2.4.7\n",
      "Collecting git+https://github.com/hms-dbmi/pic-sure-biodatacatalyst-python-adapter-hpds.git\n",
      "  Cloning https://github.com/hms-dbmi/pic-sure-biodatacatalyst-python-adapter-hpds.git to /tmp/pip-req-build-q9imwexr\n",
      "  Running command git clone -q https://github.com/hms-dbmi/pic-sure-biodatacatalyst-python-adapter-hpds.git /tmp/pip-req-build-q9imwexr\n",
      "Collecting httplib2\n",
      "  Using cached httplib2-0.19.1-py3-none-any.whl (95 kB)\n",
      "Collecting pyparsing<3,>=2.4.2\n",
      "  Using cached pyparsing-2.4.7-py2.py3-none-any.whl (67 kB)\n",
      "Building wheels for collected packages: PicSureBdcAdapter\n",
      "  Building wheel for PicSureBdcAdapter (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for PicSureBdcAdapter: filename=PicSureBdcAdapter-1.0.0-py3-none-any.whl size=8605 sha256=6698be97009b13257239041fcbe23f059ed6b11a0274eb0291b6d04d3926ef4e\n",
      "  Stored in directory: /tmp/pip-ephem-wheel-cache-s_1x1j5y/wheels/20/c3/f6/72bb8b68322e9bb56152181e827c51c09df269e5f84ed7311b\n",
      "Successfully built PicSureBdcAdapter\n",
      "Installing collected packages: pyparsing, httplib2, PicSureBdcAdapter\n",
      "  Attempting uninstall: pyparsing\n",
      "    Found existing installation: pyparsing 2.4.7\n",
      "    Uninstalling pyparsing-2.4.7:\n",
      "      Successfully uninstalled pyparsing-2.4.7\n",
      "  Attempting uninstall: httplib2\n",
      "    Found existing installation: httplib2 0.19.1\n",
      "    Uninstalling httplib2-0.19.1:\n",
      "      Successfully uninstalled httplib2-0.19.1\n",
      "Successfully installed PicSureBdcAdapter-1.0.0 httplib2-0.19.1 pyparsing-2.4.7\n"
     ]
    }
   ],
   "source": [
    "!{sys.executable} -m pip install --upgrade --force-reinstall git+https://github.com/hms-dbmi/pic-sure-python-client.git\n",
    "!{sys.executable} -m pip install --upgrade --force-reinstall git+https://github.com/hms-dbmi/pic-sure-python-adapter-hpds.git\n",
    "!{sys.executable} -m pip install --upgrade --force-reinstall git+https://github.com/hms-dbmi/pic-sure-biodatacatalyst-python-adapter-hpds.git"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "crucial-yemen",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "from pprint import pprint\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np \n",
    "import matplotlib.pyplot as plt\n",
    "from scipy import stats\n",
    "\n",
    "import PicSureClient\n",
    "import PicSureBdcAdapter\n",
    "\n",
    "from python_lib.utils import get_multiIndex_variablesDict, joining_variablesDict_onCol\n",
    "\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "wanted-charter",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pandas DataFrame display options\n",
    "pd.set_option(\"max.rows\", 100)\n",
    "\n",
    "# Matplotlib display parameters\n",
    "plt.rcParams[\"figure.figsize\"] = (14,8)\n",
    "font = {'weight' : 'bold',\n",
    "        'size'   : 12}\n",
    "plt.rc('font', **font)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "level-nudist",
   "metadata": {},
   "source": [
    "## Connecting to a PIC-SURE Network\n",
    "**Again, before running this notebook, please be sure to get a user-specific security token. For more information about how to proceed, see the \\\"Get your security token\\\" instructions in the [README.md](https://github.com/hms-dbmi/Access-to-Data-using-PIC-SURE-API/tree/harmonized_lipid_measurements_example/NHLBI_BioData_Catalyst#get-your-security-token).**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "residential-rwanda",
   "metadata": {},
   "outputs": [],
   "source": [
    "PICSURE_network_URL = \"https://picsure.biodatacatalyst.nhlbi.nih.gov/picsure\"\n",
    "resource_id = \"02e23f52-f354-4e8b-992c-d37c8b9ba140\"\n",
    "token_file = \"token.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "affecting-dependence",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(token_file, \"r\") as f:\n",
    "    my_token = f.read()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "norman-anaheim",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[38;5;91;40m\n",
      "\n",
      "+=========================================================================================+\n",
      "|        [ WARNING ] you are specifying that you WANT to allow self-signed SSL            |\n",
      "|        certificates to be acceptable for connections.  This may be useful for           |\n",
      "|        working in a development environment or on systems that host public              |\n",
      "|        data.  BEST SECURITY PRACTICES ARE THAT IF YOU ARE WORKING WITH SENSITIVE        |\n",
      "|        DATA THEN ALL SSL CERTS BY THOSE EVIRONMENTS SHOULD NOT BE SELF-SIGNED.          |\n",
      "+=========================================================================================+\n",
      "\u001b[39;49m\n",
      "+--------------------------------------+------------------------------------------------------\n",
      "|  Resource UUID                       |  Resource Name                                  \n",
      "+--------------------------------------+------------------------------------------------------\n",
      "| 02e23f52-f354-4e8b-992c-d37c8b9ba140\n",
      "+--------------------------------------+------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "client = PicSureClient.Client()\n",
    "connection = client.connect(PICSURE_network_URL, my_token, True)\n",
    "adapter = PicSureBdcAdapter.Adapter(connection)\n",
    "resource = adapter.useResource(resource_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dirty-sperm",
   "metadata": {},
   "source": [
    "## Longitudinal Lipid Variable Example\n",
    "Example showing how to extract lipid measurements from multiple visits for different cohorts\n",
    "\n",
    "### Access the data\n",
    "First, we will create a multiIndex variable dictionary of all variables we have access to."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "editorial-intelligence",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>simplified_name</th>\n",
       "      <th>name</th>\n",
       "      <th>observationCount</th>\n",
       "      <th>categorical</th>\n",
       "      <th>categoryValues</th>\n",
       "      <th>nb_modalities</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>HpdsDataType</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>level_0</th>\n",
       "      <th>level_1</th>\n",
       "      <th>level_2</th>\n",
       "      <th>level_3</th>\n",
       "      <th>level_4</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gene_with_variant</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <td>Gene_with_variant</td>\n",
       "      <td>Gene_with_variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>[HTR4, AC121758.1, HTR6, HTR7, BBX, RN7SL563P,...</td>\n",
       "      <td>39598.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>info</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Variant_class</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <td>Variant_class</td>\n",
       "      <td>Variant_class</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>[SNV, insertion, deletion]</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>info</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Variant_consequence_calculated</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <td>Variant_consequence_calculated</td>\n",
       "      <td>Variant_consequence_calculated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>[intergenic_variant, start_retained_variant, f...</td>\n",
       "      <td>28.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>info</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Variant_frequency_as_text</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <td>Variant_frequency_as_text</td>\n",
       "      <td>Variant_frequency_as_text</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>[Novel, Rare, Common]</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>info</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Variant_frequency_in_gnomAD</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <td>Variant_frequency_in_gnomAD</td>\n",
       "      <td>Variant_frequency_in_gnomAD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>info</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">_studies</th>\n",
       "      <th>NHLBI TOPMed: Whole Genome Sequencing of Venous Thromboembolism (WGS of VTE) ( phs001402 )</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <td>NHLBI TOPMed: Whole Genome Sequencing of Venou...</td>\n",
       "      <td>\\_studies\\NHLBI TOPMed: Whole Genome Sequencin...</td>\n",
       "      <td>1535.0</td>\n",
       "      <td>True</td>\n",
       "      <td>[TRUE]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>phenotypes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NHLBI TOPMed: Women's Health Initiative (WHI) ( phs001237 )</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <td>NHLBI TOPMed: Women's Health Initiative (WHI) ...</td>\n",
       "      <td>\\_studies\\NHLBI TOPMed: Women's Health Initiat...</td>\n",
       "      <td>11357.0</td>\n",
       "      <td>True</td>\n",
       "      <td>[TRUE]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>phenotypes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>The Cleveland Clinic Foundation's Lone Atrial Fibrillation GWAS Study ( phs000820 )</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <td>The Cleveland Clinic Foundation's Lone Atrial ...</td>\n",
       "      <td>\\_studies\\The Cleveland Clinic Foundation's Lo...</td>\n",
       "      <td>543.0</td>\n",
       "      <td>True</td>\n",
       "      <td>[TRUE]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>phenotypes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Women's Health Initiative Clinical Trial and Observational Study ( phs000200 )</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <td>Women's Health Initiative Clinical Trial and O...</td>\n",
       "      <td>\\_studies\\Women's Health Initiative Clinical T...</td>\n",
       "      <td>143455.0</td>\n",
       "      <td>True</td>\n",
       "      <td>[TRUE]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>phenotypes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>_topmed_consents</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <td>_topmed_consents</td>\n",
       "      <td>\\_topmed_consents\\</td>\n",
       "      <td>131940.0</td>\n",
       "      <td>True</td>\n",
       "      <td>[phs000920.c2, phs000921.c2, phs000946.c1, phs...</td>\n",
       "      <td>59.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>phenotypes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>122363 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                             simplified_name  \\\n",
       "level_0                        level_1                                            level_2 level_3 level_4                                                      \n",
       "Gene_with_variant              NaN                                                NaN     NaN     NaN                                      Gene_with_variant   \n",
       "Variant_class                  NaN                                                NaN     NaN     NaN                                          Variant_class   \n",
       "Variant_consequence_calculated NaN                                                NaN     NaN     NaN                         Variant_consequence_calculated   \n",
       "Variant_frequency_as_text      NaN                                                NaN     NaN     NaN                              Variant_frequency_as_text   \n",
       "Variant_frequency_in_gnomAD    NaN                                                NaN     NaN     NaN                            Variant_frequency_in_gnomAD   \n",
       "...                                                                                                                                                      ...   \n",
       "_studies                       NHLBI TOPMed: Whole Genome Sequencing of Venous... NaN     NaN     NaN      NHLBI TOPMed: Whole Genome Sequencing of Venou...   \n",
       "                               NHLBI TOPMed: Women's Health Initiative (WHI) (... NaN     NaN     NaN      NHLBI TOPMed: Women's Health Initiative (WHI) ...   \n",
       "                               The Cleveland Clinic Foundation's Lone Atrial F... NaN     NaN     NaN      The Cleveland Clinic Foundation's Lone Atrial ...   \n",
       "                               Women's Health Initiative Clinical Trial and Ob... NaN     NaN     NaN      Women's Health Initiative Clinical Trial and O...   \n",
       "_topmed_consents               NaN                                                NaN     NaN     NaN                                       _topmed_consents   \n",
       "\n",
       "                                                                                                                                                        name  \\\n",
       "level_0                        level_1                                            level_2 level_3 level_4                                                      \n",
       "Gene_with_variant              NaN                                                NaN     NaN     NaN                                      Gene_with_variant   \n",
       "Variant_class                  NaN                                                NaN     NaN     NaN                                          Variant_class   \n",
       "Variant_consequence_calculated NaN                                                NaN     NaN     NaN                         Variant_consequence_calculated   \n",
       "Variant_frequency_as_text      NaN                                                NaN     NaN     NaN                              Variant_frequency_as_text   \n",
       "Variant_frequency_in_gnomAD    NaN                                                NaN     NaN     NaN                            Variant_frequency_in_gnomAD   \n",
       "...                                                                                                                                                      ...   \n",
       "_studies                       NHLBI TOPMed: Whole Genome Sequencing of Venous... NaN     NaN     NaN      \\_studies\\NHLBI TOPMed: Whole Genome Sequencin...   \n",
       "                               NHLBI TOPMed: Women's Health Initiative (WHI) (... NaN     NaN     NaN      \\_studies\\NHLBI TOPMed: Women's Health Initiat...   \n",
       "                               The Cleveland Clinic Foundation's Lone Atrial F... NaN     NaN     NaN      \\_studies\\The Cleveland Clinic Foundation's Lo...   \n",
       "                               Women's Health Initiative Clinical Trial and Ob... NaN     NaN     NaN      \\_studies\\Women's Health Initiative Clinical T...   \n",
       "_topmed_consents               NaN                                                NaN     NaN     NaN                                     \\_topmed_consents\\   \n",
       "\n",
       "                                                                                                           observationCount  \\\n",
       "level_0                        level_1                                            level_2 level_3 level_4                     \n",
       "Gene_with_variant              NaN                                                NaN     NaN     NaN                   NaN   \n",
       "Variant_class                  NaN                                                NaN     NaN     NaN                   NaN   \n",
       "Variant_consequence_calculated NaN                                                NaN     NaN     NaN                   NaN   \n",
       "Variant_frequency_as_text      NaN                                                NaN     NaN     NaN                   NaN   \n",
       "Variant_frequency_in_gnomAD    NaN                                                NaN     NaN     NaN                   NaN   \n",
       "...                                                                                                                     ...   \n",
       "_studies                       NHLBI TOPMed: Whole Genome Sequencing of Venous... NaN     NaN     NaN                1535.0   \n",
       "                               NHLBI TOPMed: Women's Health Initiative (WHI) (... NaN     NaN     NaN               11357.0   \n",
       "                               The Cleveland Clinic Foundation's Lone Atrial F... NaN     NaN     NaN                 543.0   \n",
       "                               Women's Health Initiative Clinical Trial and Ob... NaN     NaN     NaN              143455.0   \n",
       "_topmed_consents               NaN                                                NaN     NaN     NaN              131940.0   \n",
       "\n",
       "                                                                                                           categorical  \\\n",
       "level_0                        level_1                                            level_2 level_3 level_4                \n",
       "Gene_with_variant              NaN                                                NaN     NaN     NaN             True   \n",
       "Variant_class                  NaN                                                NaN     NaN     NaN             True   \n",
       "Variant_consequence_calculated NaN                                                NaN     NaN     NaN             True   \n",
       "Variant_frequency_as_text      NaN                                                NaN     NaN     NaN             True   \n",
       "Variant_frequency_in_gnomAD    NaN                                                NaN     NaN     NaN            False   \n",
       "...                                                                                                                ...   \n",
       "_studies                       NHLBI TOPMed: Whole Genome Sequencing of Venous... NaN     NaN     NaN             True   \n",
       "                               NHLBI TOPMed: Women's Health Initiative (WHI) (... NaN     NaN     NaN             True   \n",
       "                               The Cleveland Clinic Foundation's Lone Atrial F... NaN     NaN     NaN             True   \n",
       "                               Women's Health Initiative Clinical Trial and Ob... NaN     NaN     NaN             True   \n",
       "_topmed_consents               NaN                                                NaN     NaN     NaN             True   \n",
       "\n",
       "                                                                                                                                              categoryValues  \\\n",
       "level_0                        level_1                                            level_2 level_3 level_4                                                      \n",
       "Gene_with_variant              NaN                                                NaN     NaN     NaN      [HTR4, AC121758.1, HTR6, HTR7, BBX, RN7SL563P,...   \n",
       "Variant_class                  NaN                                                NaN     NaN     NaN                             [SNV, insertion, deletion]   \n",
       "Variant_consequence_calculated NaN                                                NaN     NaN     NaN      [intergenic_variant, start_retained_variant, f...   \n",
       "Variant_frequency_as_text      NaN                                                NaN     NaN     NaN                                  [Novel, Rare, Common]   \n",
       "Variant_frequency_in_gnomAD    NaN                                                NaN     NaN     NaN                                                     []   \n",
       "...                                                                                                                                                      ...   \n",
       "_studies                       NHLBI TOPMed: Whole Genome Sequencing of Venous... NaN     NaN     NaN                                                 [TRUE]   \n",
       "                               NHLBI TOPMed: Women's Health Initiative (WHI) (... NaN     NaN     NaN                                                 [TRUE]   \n",
       "                               The Cleveland Clinic Foundation's Lone Atrial F... NaN     NaN     NaN                                                 [TRUE]   \n",
       "                               Women's Health Initiative Clinical Trial and Ob... NaN     NaN     NaN                                                 [TRUE]   \n",
       "_topmed_consents               NaN                                                NaN     NaN     NaN      [phs000920.c2, phs000921.c2, phs000946.c1, phs...   \n",
       "\n",
       "                                                                                                           nb_modalities  \\\n",
       "level_0                        level_1                                            level_2 level_3 level_4                  \n",
       "Gene_with_variant              NaN                                                NaN     NaN     NaN            39598.0   \n",
       "Variant_class                  NaN                                                NaN     NaN     NaN                3.0   \n",
       "Variant_consequence_calculated NaN                                                NaN     NaN     NaN               28.0   \n",
       "Variant_frequency_as_text      NaN                                                NaN     NaN     NaN                3.0   \n",
       "Variant_frequency_in_gnomAD    NaN                                                NaN     NaN     NaN                0.0   \n",
       "...                                                                                                                  ...   \n",
       "_studies                       NHLBI TOPMed: Whole Genome Sequencing of Venous... NaN     NaN     NaN                1.0   \n",
       "                               NHLBI TOPMed: Women's Health Initiative (WHI) (... NaN     NaN     NaN                1.0   \n",
       "                               The Cleveland Clinic Foundation's Lone Atrial F... NaN     NaN     NaN                1.0   \n",
       "                               Women's Health Initiative Clinical Trial and Ob... NaN     NaN     NaN                1.0   \n",
       "_topmed_consents               NaN                                                NaN     NaN     NaN               59.0   \n",
       "\n",
       "                                                                                                           min  \\\n",
       "level_0                        level_1                                            level_2 level_3 level_4        \n",
       "Gene_with_variant              NaN                                                NaN     NaN     NaN      NaN   \n",
       "Variant_class                  NaN                                                NaN     NaN     NaN      NaN   \n",
       "Variant_consequence_calculated NaN                                                NaN     NaN     NaN      NaN   \n",
       "Variant_frequency_as_text      NaN                                                NaN     NaN     NaN      NaN   \n",
       "Variant_frequency_in_gnomAD    NaN                                                NaN     NaN     NaN      NaN   \n",
       "...                                                                                                        ...   \n",
       "_studies                       NHLBI TOPMed: Whole Genome Sequencing of Venous... NaN     NaN     NaN      NaN   \n",
       "                               NHLBI TOPMed: Women's Health Initiative (WHI) (... NaN     NaN     NaN      NaN   \n",
       "                               The Cleveland Clinic Foundation's Lone Atrial F... NaN     NaN     NaN      NaN   \n",
       "                               Women's Health Initiative Clinical Trial and Ob... NaN     NaN     NaN      NaN   \n",
       "_topmed_consents               NaN                                                NaN     NaN     NaN      NaN   \n",
       "\n",
       "                                                                                                           max  \\\n",
       "level_0                        level_1                                            level_2 level_3 level_4        \n",
       "Gene_with_variant              NaN                                                NaN     NaN     NaN      NaN   \n",
       "Variant_class                  NaN                                                NaN     NaN     NaN      NaN   \n",
       "Variant_consequence_calculated NaN                                                NaN     NaN     NaN      NaN   \n",
       "Variant_frequency_as_text      NaN                                                NaN     NaN     NaN      NaN   \n",
       "Variant_frequency_in_gnomAD    NaN                                                NaN     NaN     NaN      NaN   \n",
       "...                                                                                                        ...   \n",
       "_studies                       NHLBI TOPMed: Whole Genome Sequencing of Venous... NaN     NaN     NaN      NaN   \n",
       "                               NHLBI TOPMed: Women's Health Initiative (WHI) (... NaN     NaN     NaN      NaN   \n",
       "                               The Cleveland Clinic Foundation's Lone Atrial F... NaN     NaN     NaN      NaN   \n",
       "                               Women's Health Initiative Clinical Trial and Ob... NaN     NaN     NaN      NaN   \n",
       "_topmed_consents               NaN                                                NaN     NaN     NaN      NaN   \n",
       "\n",
       "                                                                                                          HpdsDataType  \n",
       "level_0                        level_1                                            level_2 level_3 level_4               \n",
       "Gene_with_variant              NaN                                                NaN     NaN     NaN             info  \n",
       "Variant_class                  NaN                                                NaN     NaN     NaN             info  \n",
       "Variant_consequence_calculated NaN                                                NaN     NaN     NaN             info  \n",
       "Variant_frequency_as_text      NaN                                                NaN     NaN     NaN             info  \n",
       "Variant_frequency_in_gnomAD    NaN                                                NaN     NaN     NaN             info  \n",
       "...                                                                                                                ...  \n",
       "_studies                       NHLBI TOPMed: Whole Genome Sequencing of Venous... NaN     NaN     NaN       phenotypes  \n",
       "                               NHLBI TOPMed: Women's Health Initiative (WHI) (... NaN     NaN     NaN       phenotypes  \n",
       "                               The Cleveland Clinic Foundation's Lone Atrial F... NaN     NaN     NaN       phenotypes  \n",
       "                               Women's Health Initiative Clinical Trial and Ob... NaN     NaN     NaN       phenotypes  \n",
       "_topmed_consents               NaN                                                NaN     NaN     NaN       phenotypes  \n",
       "\n",
       "[122363 rows x 9 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fullVariableDict = resource.dictionary().find().DataFrame()\n",
    "variablesDict = get_multiIndex_variablesDict(fullVariableDict)\n",
    "variablesDict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "minute-lemon",
   "metadata": {},
   "source": [
    "In this example, we are interested in variables related to lipids. We can find all variables related to the search terms 'lipid' and 'triglyceride' through applying the following filter on the multiIndex dictionary:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "generic-director",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(24, 9)"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mask_lipid = [type(i) == str and \"lipid\" in i for i in variablesDict.index.get_level_values(2)]\n",
    "mask_triglyceride = [type(i) == str and \"triglyceride\" in i for i in variablesDict.index.get_level_values(2)]\n",
    "\n",
    "lipid_vars = variablesDict.loc[mask_lipid or mask_triglyceride,:]\n",
    "\n",
    "lipid_vars.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "million-parcel",
   "metadata": {},
   "source": [
    "### Identify the longitudinal lipid variables\n",
    "This block of code does the following:\n",
    "\n",
    "- uses the multiindex dataframe containing variables which are related to 'lipid' or 'triglyceride'\n",
    "- filters for variables with keywords 'exam #' or 'visit #'\n",
    "- extracts the exam number of each variable into column exam_number\n",
    "- groups variables by study (level_0) and longitudinal variable (longvar)\n",
    "- returns a table showing the variables that have more than one exam recorded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "integrated-shipping",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "results at ExAm 43 lipids\n"
     ]
    }
   ],
   "source": [
    "test = \"results at ExAm 43 lipids\"\n",
    "print(re.search('.* (exam|visit) \\d+ .*', test, re.IGNORECASE).group())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "hairy-color",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>simplified_name</th>\n",
       "      <th>name</th>\n",
       "      <th>observationCount</th>\n",
       "      <th>categorical</th>\n",
       "      <th>categoryValues</th>\n",
       "      <th>nb_modalities</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>HpdsDataType</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>level_0</th>\n",
       "      <th>level_1</th>\n",
       "      <th>level_2</th>\n",
       "      <th>level_3</th>\n",
       "      <th>level_4</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [simplified_name, name, observationCount, categorical, categoryValues, nb_modalities, min, max, HpdsDataType]\n",
       "Index: []"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#re.search('.* (exam|visit) \\d+ .*', lipid_vars[\"name\"], re.IGNORECASE).group()\n",
    "#lipid_vars[lipid_vars['name'].str.contains('.* (exam|visit) \\d+ .*')==True]\n",
    "#longitudinal_lipid_vars = lipid_vars[lipid_vars[\"name\"].str.contains('(exam|visit) \\d+')==True]\n",
    "#longitudinal_lipid_vars.shape\n",
    "lipid_vars[lipid_vars[\"name\"].str.contains('visit \\d+')==True]# | lipid_vars[\"name\"].str.contains('visit \\d+')==True)]"
   ]
  },
  {
   "cell_type": "raw",
   "id": "assured-inquiry",
   "metadata": {},
   "source": [
    "longitudinal_lipid_vars <- lipid_vars %>% \n",
    "\n",
    "  # filter for longitudinal varibles; here defined as variable names containing 'exam' or 'visit' (case insensitive)\n",
    "  filter((grepl('Exam \\\\d+', name, ignore.case = TRUE) | \n",
    "         grepl('Visit \\\\d+', name, ignore.case = TRUE))) %>%\n",
    "\n",
    "  # extract the exam / visit number and store it as a new variable: 'exam_number'\n",
    "  mutate(exam_number = str_extract(name, regex(\"(exam \\\\d+)|(visit \\\\d+)\", ignore_case=T)),\n",
    "         \n",
    "         # extract the longitudinal variable name by removing the exam / visit number and store as new variable: 'longvar'\n",
    "         longvar =  str_replace(name, regex('(exam \\\\d+.$)|(visit \\\\d+.$)', ignore_case = T), '')) %>%\n",
    "\n",
    "  # group our variable data by study name (level_0) and longitudinal variable name (longvar)\n",
    "  group_by(level_0, longvar) %>%\n",
    "\n",
    "  # count number of exams / visits within each distinct study - variable pairing and store as new variable: 'n_exams'\n",
    "  summarise(n_exams = n_distinct(exam_number)) %>% \n",
    "\n",
    "  # filter results to only include variables which have more than one exam / visit (n_exams > 1)\n",
    "  filter(n_exams > 1) %>%\n",
    "\n",
    "  # sort results by number of exams / visits \n",
    "  arrange(desc(n_exams))\n",
    "\n",
    "\n",
    "longitudinal_lipid_vars"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "conda_python3",
   "language": "python",
   "name": "conda_python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
